---
title: "Interpretation of GSFA Results on LUHMES CROP-seq Data"
subtitle: "-- for manuscript"
author: "Yifan Zhou (zhouyf@uchicago.edu)"
output:
  html_document:
    number_sections: yes
    toc: yes
    toc_float: true
---

```{r setup, include=FALSE}
requireNamespace("pander", quietly = TRUE)
# set default chunk output
knitr::opts_chunk$set(echo = T,
                      message = FALSE,
                      warning = FALSE,
                      comment = NA,
                      tidy = FALSE,
                      fig.width = 8,
                      fig.height = 6,
                      fig.align = "center",
                      results = "asis")

# formatting of pander tables
pander::panderOptions('knitr.auto.asis', FALSE)
pander::panderOptions("table.split.table", Inf)
```

# Introduction

This tutorial demonstrates how to visualize and interpret the results from a GSFA run.

We have described how to run GSFA on LUHMES CROP-seq data from Lalli et al. [here](https://gradonion.github.io/GSFA_paper/preprocess_and_gsfa_LUHMES.html).

To recapitulate, the processed dataset consists of 8708 neural progenitor cells that belong to one of the 15 perturbation conditions (CRISPR knock-down of 14 neurodevelopmental genes, and negative control). Top 6000 genes ranked by deviance statistics were kept. And GSFA was performed on the data with 20 factors specified.

## Load necessary packages and data

```{r}
library(data.table)
library(Matrix)
library(tidyverse)
library(ggplot2)
theme_set(theme_bw() + theme(plot.title = element_text(size = 14, hjust = 0.5),
                             axis.title = element_text(size = 14),
                             axis.text = element_text(size = 12),
                             legend.title = element_text(size = 13),
                             legend.text = element_text(size = 12),
                             panel.grid.minor = element_blank())
)
library(gridExtra)
library(ComplexHeatmap)
library(kableExtra)
library(WebGestaltR)

source("../R/plotting_functions.R")
data_folder <- "../data/LUHMES/"
```

The first thing we need is the output of GSFA `fit_gsfa_multivar()` run. The lighter version containing just the posterior mean estimates and LFSR of perturbation-gene effects is enough. (See `R/run_gsfa_LUHMES.R` for more GSFA run details.)

```{r}
fit <- readRDS(paste0(data_folder, "gsfa_fit.light.rds"))
gibbs_PM <- fit$posterior_means
lfsr_mat <- fit$lfsr[, -ncol(fit$lfsr)]
KO_names <- colnames(lfsr_mat)
```

We also need the cell by perturbation matrix which was used as input $G$ for GSFA.

```{r}
G_mat <- readRDS(paste0(data_folder, "perturbation_matrix.rds"))
```

Finally, we load the mapping from gene name to ENSEMBL ID for all 6k genes used in GSFA, as well as selected neuronal marker genes. This is specific to this study and analysis.

```{r}
genes_df <- readRDS(paste0(data_folder, "top6k_genes.rds"))
interest_df <- readRDS(paste0(data_folder, "selected_neuronal_markers.rds"))
```

# Factor ~ Perturbation Association

## Perturbation effects on factors

Fisrt of all, we look at the estimated effects of gene perturbations on factors inferred by GSFA. 

We found that targeting of 7 genes, ADNP, ARID1B, ASH1L, CHD2, DYRK1A, PTEN, and SETD5, has significant effects (PIP > 0.95) on at least 1 of the 20 inferred factors.

All targets and factors:
```{r, fig.width=9, fig.height=5.5}
dotplot_beta_PIP(t(gibbs_PM$Gamma_pm), t(gibbs_PM$beta_pm),
                 marker_names = KO_names,
                 reorder_markers = c(KO_names[KO_names!="Nontargeting"], "Nontargeting"),
                 inverse_factors = F) +
  coord_flip()
```

```{r fig.width=8, fig.height=5.5, eval=F}
## Similar visualization using GSFA built-in functions:
GSFA::dotplot_beta_PIP(fit,
                       target_names = KO_names,
                       reorder_targets = c(KO_names[KO_names!="Nontargeting"], "Nontargeting"))
```

Here is a closer look at the estimated effects of selected perturbations on selected factors:

```{r, fig.width=4, fig.height=4}
targets <- c("ADNP", "ARID1B", "ASH1L", "CHD2", "DYRK1A", "PTEN", "SETD5")
complexplot_perturbation_factor(gibbs_PM$Gamma_pm[-nrow(gibbs_PM$Gamma_pm), ],
                                gibbs_PM$beta_pm[-nrow(gibbs_PM$beta_pm), ],
                                marker_names = KO_names,
                                reorder_markers = targets,
                                reorder_factors = c(4, 9, 16))
```

```{r fig.width=4, fig.height=4, eval=F}
## Similar visualization using GSFA built-in functions:
GSFA::dotplot_beta_PIP(fit,
                       target_names = KO_names,
                       reorder_targets = targets, reorder_factors = c(4, 9, 16))
```

## Factor-perturbation association p values

We can also assess the correlations between each pair of perturbation and inferred factor.  
The distribution of correlation p values show significant signals.

```{r fig.width=10, fig.height=5}
gibbs_res_tb <- make_gibbs_res_tb(gibbs_PM, G_mat, compute_pve = F)
heatmap_matrix <- gibbs_res_tb %>% select(starts_with("pval"))
rownames(heatmap_matrix) <- 1:nrow(heatmap_matrix)
colnames(heatmap_matrix) <- colnames(G_mat)

summ_pvalues(unlist(heatmap_matrix),
             title_text = "GSFA\n(15 Targets x 20 Factors)")
```

# Factor Interpretation

## Correlation within factors

Since the GSFA model does not enforce orthogonality among factors, we first inspect the pairwise correlation within them to see if there is any redundancy. As we can see below, the inferred factors are mostly independent of each other.

```{r fig.width=7, fig.height=5}
plot_pairwise.corr_heatmap(input_mat_1 = gibbs_PM$Z_pm,
                           corr_type = "pearson",
                           name_1 = "Pairwise correlation within factors (Z)",
                           label_size = 10)
```

```{r fig.width=7, fig.height=5}
plot_pairwise.corr_heatmap(input_mat_1 = (gibbs_PM$F_pm > 0.95) * 1,
                           corr_type = "jaccard",
                           name_1 = "Pairwise correlation within \nbinarized gene loadings (F_pm > 0.95)",
                           label_size = 10)
```

## Gene loading in factors

To understand these latent factors, we inspect the loadings (weights) of several marker genes for neuron maturation and differentiation in them.

```{r echo=FALSE}
interest_df <- interest_df[interest_df$gene_name %in% genes_df$Name, ]
rownames(interest_df) <- NULL
knitr::kable(interest_df) %>%
    kable_styling() %>% scroll_box(width = '100%', height = '400px')
```

We visualize both the gene PIPs (dot size) and gene weights (dot color) in all factors:

```{r, fig.width=10, fig.height=5}
complexplot_gene_factor(genes_df, interest_df, gibbs_PM$F_pm, gibbs_PM$W_pm)
```

A closer look at some factors that are associated with perturbations:

```{r, fig.width=6.5, fig.height=5}
complexplot_gene_factor(genes_df, interest_df, gibbs_PM$F_pm, gibbs_PM$W_pm,
                        reorder_factors = c(4, 9, 16))
```

## GO enrichment analysis in factors

To further characterize these latent factors, we perform GO (gene ontology) enrichment analysis of genes loaded on the factors using `WebGestalt`.

Foreground genes: genes w/ non-zero loadings in each factor (gene PIP > 0.95);    
Background genes: all 6000 genes used in GSFA;    
Statistical test: hypergeometric test (over-representation test);  
Gene sets: GO Slim "Biological Process" (non-redundant).

```{r eval=FALSE}
## The "WebGestaltR" tool needs Internet connection.
enrich_db <- "geneontology_Biological_Process_noRedundant"
PIP_mat <- gibbs_PM$F_pm
enrich_res_by_factor <- list()
for (i in 1:ncol(PIP_mat)){
  enrich_res_by_factor[[i]] <- 
    WebGestaltR::WebGestaltR(enrichMethod = "ORA",
                             organism = "hsapiens",
                             enrichDatabase = enrich_db,
                             interestGene = genes_df[PIP_mat[, i] > 0.95, ]$ID,
                             interestGeneType = "ensembl_gene_id",
                             referenceGene = genes_df$ID,
                             referenceGeneType = "ensembl_gene_id",
                             isOutput = F)
}
```

```{r echo=FALSE}
enrich_res_by_factor <- readRDS(paste0(data_folder, "GO_enrich_in_GSFA_factors.rds"))
```

Several GO “biological process” terms related to neuronal development are enriched in factors 4, 9, and 16:

```{r, fig.height=2.4, fig.width=7}
factor_indx <- 4
terms_of_interest <- c("regulation of ion transmembrane transport",
                       "regulation of trans-synaptic signaling",
                       "axon development",
                       "regulation of neuron projection development")
barplot_top_enrich_terms(enrich_res_by_factor[[factor_indx]],
                         terms_of_interest = terms_of_interest,
                         str_wrap_length = 35, pval_max = 8, FC_max = 6) +
  labs(title = paste0("Factor ", factor_indx),
       x = "Fold of enrichment")
```

```{r, fig.height=2.4, fig.width=7}
factor_indx <- 9
terms_of_interest <- c("actin filament organization",
                       "cell fate commitment",
                       "regulation of neuron projection development",
                       "regulation of cell morphogenesis")
barplot_top_enrich_terms(enrich_res_by_factor[[factor_indx]],
                         terms_of_interest = terms_of_interest,
                         str_wrap_length = 35, pval_max = 8, FC_max = 6) +
  labs(title = paste0("Factor ", factor_indx),
       x = "Fold of enrichment")
```

```{r, fig.height=1.8, fig.width=7}
factor_indx <- 16
terms_of_interest <- c("developmental growth involved in morphogenesis",
                       "axon development")
barplot_top_enrich_terms(enrich_res_by_factor[[factor_indx]],
                         terms_of_interest = terms_of_interest, 
                         str_wrap_length = 35, pval_max = 8, FC_max = 6) +
  labs(title = paste0("Factor ", factor_indx),
       x = "Fold of enrichment")
```

# DEG Interpretation

In GSFA, differential expression analysis can be performed based on the LFSR method. Here we evaluate the specific downstream genes affected by the perturbations detected by GSFA.

We also performed several other differential expression methods for comparison, including scMAGeCK-LR, MAST, and DESeq.

## Number of DEGs detected by different methods

```{r}
fdr_cutoff <- 0.05
lfsr_cutoff <- 0.05
```

```{r echo=FALSE}
lfsr_signif_num <- colSums(lfsr_mat < lfsr_cutoff)
signif_num_tb <- t(data.frame(KO = names(lfsr_signif_num),
                           Num_genes = lfsr_signif_num,
                           row.names = NULL))
knitr::kable(rbind(signif_num_tb[, 1:5], signif_num_tb[, 6:10], signif_num_tb[, 11:15]),
             caption = "Number of DEGs detected by GSFA under each perturbation:") %>%
  kable_styling() %>% scroll_box(width = '100%')
```

```{r deseq}
deseq_list <- readRDS(paste0(data_folder, "DE_result_DESeq2.rds"))
deseq_signif_counts <- sapply(deseq_list, function(x){filter(x, FDR < fdr_cutoff) %>% nrow()})
```

```{r mast}
mast_list <- readRDS(paste0(data_folder, "DE_result_MAST.rds"))
mast_signif_counts <- sapply(mast_list, function(x){filter(x, FDR < fdr_cutoff) %>% nrow()})
```

```{r}
scmageck_res <- readRDS(paste0(data_folder, "DE_result_scMAGeCK.rds"))
colnames(scmageck_res$fdr)[colnames(scmageck_res$fdr) == "NegCtrl"] <- "Nontargeting"
scmageck_signif_counts <- colSums(scmageck_res$fdr[, KO_names] < fdr_cutoff)
```

```{r dge_comparison_merge}
dge_comparison_df <- data.frame(Perturbation = names(lfsr_signif_num),
                                GSFA = lfsr_signif_num,
                                scMAGeCK = scmageck_signif_counts,
                                DESeq2 = deseq_signif_counts,
                                MAST = mast_signif_counts)
dge_comparison_df$Perturbation[dge_comparison_df$Perturbation == "Nontargeting"] <- "NegCtrl"
```

Number of DEGs detected under each perturbation using 4 different methods:

```{r, fig.width=13, fig.height=5}
dge_plot_df <- reshape2::melt(dge_comparison_df, id.var = "Perturbation",
                              variable.name = "Method", value.name = "Num_DEGs")
dge_plot_df$Perturbation <- factor(dge_plot_df$Perturbation,
                                   levels = c("NegCtrl", KO_names[KO_names!="Nontargeting"]))
ggplot(dge_plot_df, aes(x = Perturbation, y = Num_DEGs+1, fill = Method)) +
  geom_bar(position = "dodge", stat = "identity") +
  geom_text(aes(label = Num_DEGs), position=position_dodge(width=0.9), vjust=-0.25) +
  scale_y_log10() +
  scale_fill_brewer(palette = "Set2") +
  labs(x = "Target gene",
       y = "Number of DEGs",
       title = "Number of DEGs detected by different methods") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 12),
        legend.position = "bottom",
        legend.text = element_text(size = 13))
```

Compared with other differential expression analysis methods, GSFA detected the most DEGs for 5 out of 7 gene targets that have significant effects.

## Perturbation effects on marker genes

To better understand the functions of these 7 target genes, we examined their effects on marker genes for neuron maturation and differentiation.

### GSFA

Here are the summarized effects of perturbations on marker genes estimated by GSFA.

As we can see, knockdown of ADNP, ASH1L, CHD2, and DYRK1A has mostly negative effects on mature neuronal markers, and positive effects on negative regulators of neuron projection, indicating delayed neuron maturation.

Knockdown of PTEN and SETD5 has the opposite pattern, which indicates accelerated neuron maturation.

```{r, fig.width=6.5, fig.height=5}
targets <- c("ADNP", "ARID1B", "ASH1L", "CHD2", "DYRK1A", "PTEN", "SETD5")
complexplot_gene_perturbation(genes_df, interest_df,
                              targets = targets,
                              lfsr_mat = lfsr_mat,
                              effect_mat = gibbs_PM$W_pm %*%
                                t(gibbs_PM$beta_pm[-nrow(gibbs_PM$beta_pm), ]))
```

```{r fig.width=5.5, fig.height=5, eval=F}
## Similar visualization using GSFA built-in functions:
GSFA::dotplot_total_effect(fit,
                           gene_indices = match(interest_df$gene_ID, rownames(lfsr_mat)),
                           gene_names = interest_df$gene_name,
                           reorder_targets = targets, 
                           plot_max_score = 0.2)
```

### scMAGeCK

Here are scMAGeCK estimated effects of perturbations on marker genes:

```{r Fig_5H, fig.width=6.5, fig.height=5}
score_mat <- scmageck_res$score
fdr_mat <- scmageck_res$fdr
complexplot_gene_perturbation(genes_df, interest_df,
                              targets = targets,
                              lfsr_mat = fdr_mat, lfsr_name = "FDR",
                              effect_mat = score_mat, effect_name = "scMAGeCK\nselection score", 
                              score_break = c(-0.2, 0, 0.2),
                              color_break = c("blue3", "grey90", "red3"))
```

### DESeq2

```{r}
FC_mat <- matrix(nrow = nrow(interest_df), ncol = length(targets))
rownames(FC_mat) <- interest_df$gene_name
colnames(FC_mat) <- targets
fdr_mat <- FC_mat
for (m in targets){
  FC_mat[, m] <- deseq_list[[m]]$log2FoldChange[match(interest_df$gene_ID,
                                                      deseq_list[[m]]$geneID)]
  fdr_mat[, m] <- deseq_list[[m]]$FDR[match(interest_df$gene_ID, deseq_list[[m]]$geneID)]
}
```

Here are DESeq2 estimated effects of perturbations on marker genes:

```{r Fig_S5C, fig.width=6.5, fig.height=5}
complexplot_gene_perturbation(genes_df, interest_df,
                              targets = targets,
                              lfsr_mat = fdr_mat, lfsr_name = "FDR",
                              effect_mat = FC_mat, effect_name = "DESeq2 log2FC",
                              score_break = c(-0.4, 0, 0.4),
                              color_break = c("blue3", "grey90", "red3"))
```

### MAST

```{r}
FC_mat <- matrix(nrow = nrow(interest_df), ncol = length(targets))
rownames(FC_mat) <- interest_df$gene_name
colnames(FC_mat) <- targets
fdr_mat <- FC_mat
for (m in targets){
  FC_mat[, m] <- mast_list[[m]]$logFC[match(interest_df$gene_ID, mast_list[[m]]$geneID)]
  fdr_mat[, m] <- mast_list[[m]]$FDR[match(interest_df$gene_ID, mast_list[[m]]$geneID)]
}
```

MAST estimated effects of perturbations on marker genes:

```{r Fig_S5D, fig.width=6.5, fig.height=5}
complexplot_gene_perturbation(genes_df, interest_df,
                              targets = targets,
                              lfsr_mat = fdr_mat, lfsr_name = "FDR",
                              effect_mat = FC_mat, effect_name = "MAST logFC",
                              score_break = c(-0.4, 0, 0.4),
                              color_break = c("blue3", "grey90", "red3"))
```

## GO enrichment in DEGs

We further examine these DEGs for enrichment of relevant biological processes through GO enrichment analysis.

Foreground genes: Genes w/ GSFA LFSR < 0.05 under each perturbation;    
Background genes: all 6000 genes used in GSFA;    
Statistical test: hypergeometric test (over-representation test);    
Gene sets: Gene ontology "Biological Process" (non-redundant).

```{r eval=FALSE}
## The "WebGestaltR" tool needs Internet connection.
targets <- names(lfsr_signif_num)[lfsr_signif_num > 0]
enrich_db <- "geneontology_Biological_Process_noRedundant"
enrich_res <- list()
for (i in targets){
  print(i)
  interest_genes <- genes_df %>% mutate(lfsr = lfsr_mat[, i]) %>%
    filter(lfsr < lfsr_cutoff) %>% pull(ID)
  enrich_res[[i]] <- 
    WebGestaltR::WebGestaltR(enrichMethod = "ORA",
                             organism = "hsapiens",
                             enrichDatabase = enrich_db,
                             interestGene = interest_genes,
                             interestGeneType = "ensembl_gene_id",
                             referenceGene = genes_df$ID,
                             referenceGeneType = "ensembl_gene_id",
                             isOutput = F)
}
```

```{r echo=FALSE}
enrich_res <- readRDS(paste0(data_folder, "GO_enrich_in_GSFA_DEGs.rds"))
```

```{r}
signif_GO_list <- list()
for (i in names(enrich_res)) {
  signif_GO_list[[i]] <- enrich_res[[i]] %>%
    dplyr::filter(FDR < 0.05) %>%
    dplyr::select(geneSet, description, size, enrichmentRatio, pValue) %>%
    mutate(target = i)
}
signif_term_df <- do.call(rbind, signif_GO_list) %>%
  group_by(geneSet, description, size) %>%
  summarise(pValue = min(pValue)) %>%
  ungroup()

abs_FC_colormap <- circlize::colorRamp2(breaks = c(0, 3, 6),
                                        colors = c("grey95", "#77d183", "#255566"))
```

```{r}
targets <- names(enrich_res)
enrich_table <- data.frame(matrix(nrow = nrow(signif_term_df),
                                  ncol = length(targets)),
                           row.names = signif_term_df$geneSet)
colnames(enrich_table) <- targets
for (i in 1:ncol(enrich_table)){
  m <- colnames(enrich_table)[i]
  enrich_df <- enrich_res[[m]] %>% dplyr::filter(FDR < 0.05)
  enrich_table[enrich_df$geneSet, i] <- enrich_df$enrichmentRatio
}
rownames(enrich_table) <- signif_term_df$description
```

```{r echo=FALSE}
terms_of_interest <- list()
terms_of_interest[[1]] <- 
  data.frame(description = c("axon development",
                             "central nervous system neuron differentiation",
                             "neuron projection guidance",
                             "regulation of neuron projection development"))

terms_of_interest[[2]] <- 
  data.frame(description = c("regulation of transporter activity",
                             "regulation of membrane potential",
                             "regulation of ion transmembrane transport",
                             "regulation of peptide secretion",
                             "signal release"))

terms_of_interest[[3]] <- 
  data.frame(description = c("neuropeptide signaling pathway",
                             "cell communication by electrical coupling",
                             "regulation of trans-synaptic signaling",
                             "regulation of neurological system process",
                             "neural precursor cell proliferation"))

terms_of_interest[[4]] <- 
  data.frame(description = c("RNA catabolic process",
                             "ribonucleotide metabolic process",
                             "ribonucleoprotein complex biogenesis",
                             "mitochondrial respiratory chain complex assembly",
                             "response to purine-containing compound"))

terms_of_interest_df <- do.call(rbind, terms_of_interest)
terms_of_interest_df <- left_join(terms_of_interest_df, signif_term_df,
                                  by = "description")
```

Here are selected GO "biological process" terms and their folds of enrichment in DEGs detected by GSFA:  
(In the code below, we omitted the content in `terms_of_interest_df` as one can subset the `enrich_table` with any terms of their choice.)

```{r, fig.height=5.5, fig.width=8}
interest_enrich_table <- enrich_table[terms_of_interest_df$description,
                                      colnames(enrich_table) != "DYRK1A"]
interest_enrich_table[is.na(interest_enrich_table)] <- 0

map <- Heatmap(abs(interest_enrich_table),
               name = "Fold of enrichment",
               col = abs_FC_colormap,
               na_col = "grey90",
               row_title = NULL, column_title = NULL,
               cluster_rows = F, cluster_columns = F,
               show_row_dend = F, show_column_dend = F,
               show_heatmap_legend = T,
               row_names_gp = gpar(fontsize = 10.5),
               column_names_rot = 45,
               column_names_side = "top",
               width = unit(6, "cm"))
draw(map, heatmap_legend_side = "bottom")
```

# Session Information

```{r results="markup"}
sessionInfo()
```
