<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Yifan Zhou (zhouyf@uchicago.edu)" />


<title>Interpretation of GSFA Results on LUHMES CROP-seq Data</title>

<script src="site_libs/jquery-3.5.1/jquery-3.5.1.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<script src="site_libs/kePrint-0.0.1/kePrint.js"></script>
<link href="site_libs/lightable-0.0.1/lightable.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">GSFA</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Interpretation of GSFA Results on LUHMES CROP-seq Data</h1>
<h3 class="subtitle">-- for manuscript</h3>
<h4 class="author">Yifan Zhou (<a href="mailto:zhouyf@uchicago.edu">zhouyf@uchicago.edu</a>)</h4>

</div>


<div id="introduction" class="section level1">
<h1><span class="header-section-number">1</span> Introduction</h1>
<p>This tutorial demonstrates how to visualize and interpret the results from a GSFA run.</p>
<p>We have described how to run GSFA on LUHMES CROP-seq data from Lalli et al. <a href="https://gradonion.github.io/GSFA_paper/preprocess_and_gsfa_LUHMES.html">here</a>.</p>
<p>To recapitulate, the processed dataset consists of 8708 neural progenitor cells that belong to one of the 15 perturbation conditions (CRISPR knock-down of 14 neurodevelopmental genes, and negative control). Top 6000 genes ranked by deviance statistics were kept. And GSFA was performed on the data with 20 factors specified.</p>
<div id="load-necessary-packages-and-data" class="section level2">
<h2><span class="header-section-number">1.1</span> Load necessary packages and data</h2>
<pre class="r"><code>library(data.table)
library(Matrix)
library(tidyverse)
library(ggplot2)
theme_set(theme_bw() + theme(plot.title = element_text(size = 14, hjust = 0.5),
                             axis.title = element_text(size = 14),
                             axis.text = element_text(size = 12),
                             legend.title = element_text(size = 13),
                             legend.text = element_text(size = 12),
                             panel.grid.minor = element_blank())
)
library(gridExtra)
library(ComplexHeatmap)
library(kableExtra)
library(WebGestaltR)

source(&quot;../R/plotting_functions.R&quot;)
data_folder &lt;- &quot;../data/LUHMES/&quot;</code></pre>
<p>The first thing we need is the output of GSFA <code>fit_gsfa_multivar()</code> run. The lighter version containing just the posterior mean estimates and LFSR of perturbation-gene effects is enough. (See <code>R/run_gsfa_LUHMES.R</code> for more GSFA run details.)</p>
<pre class="r"><code>fit &lt;- readRDS(paste0(data_folder, &quot;gsfa_fit.light.rds&quot;))
gibbs_PM &lt;- fit$posterior_means
lfsr_mat &lt;- fit$lfsr[, -ncol(fit$lfsr)]
KO_names &lt;- colnames(lfsr_mat)</code></pre>
<p>We also need the cell by perturbation matrix which was used as input <span class="math inline">\(G\)</span> for GSFA.</p>
<pre class="r"><code>G_mat &lt;- readRDS(paste0(data_folder, &quot;perturbation_matrix.rds&quot;))</code></pre>
<p>Finally, we load the mapping from gene name to ENSEMBL ID for all 6k genes used in GSFA, as well as selected neuronal marker genes. This is specific to this study and analysis.</p>
<pre class="r"><code>genes_df &lt;- readRDS(paste0(data_folder, &quot;top6k_genes.rds&quot;))
interest_df &lt;- readRDS(paste0(data_folder, &quot;selected_neuronal_markers.rds&quot;))</code></pre>
</div>
</div>
<div id="factor-perturbation-association" class="section level1">
<h1><span class="header-section-number">2</span> Factor ~ Perturbation Association</h1>
<div id="perturbation-effects-on-factors" class="section level2">
<h2><span class="header-section-number">2.1</span> Perturbation effects on factors</h2>
<p>Fisrt of all, we look at the estimated effects of gene perturbations on factors inferred by GSFA.</p>
<p>We found that targeting of 7 genes, ADNP, ARID1B, ASH1L, CHD2, DYRK1A, PTEN, and SETD5, has significant effects (PIP &gt; 0.95) on at least 1 of the 20 inferred factors.</p>
<p>All targets and factors (Figure S6A):</p>
<pre class="r"><code>dotplot_beta_PIP(t(gibbs_PM$Gamma_pm), t(gibbs_PM$beta_pm),
                 marker_names = KO_names,
                 reorder_markers = c(KO_names[KO_names!=&quot;Nontargeting&quot;], &quot;Nontargeting&quot;),
                 inverse_factors = F) +
  coord_flip()</code></pre>
<p><img src="gsfa_result_interpret_LUHMES_files/figure-html/Fig_S6A-1.png" width="864" style="display: block; margin: auto;" /></p>
<pre class="r"><code>## Similar visualization using GSFA built-in functions:
GSFA::dotplot_beta_PIP(fit,
                       target_names = KO_names,
                       reorder_targets = c(KO_names[KO_names!=&quot;Nontargeting&quot;], &quot;Nontargeting&quot;))</code></pre>
<p>Here is a closer look at the estimated effects of selected perturbations on selected factors (Figure 5A):</p>
<pre class="r"><code>targets &lt;- c(&quot;ADNP&quot;, &quot;ARID1B&quot;, &quot;ASH1L&quot;, &quot;CHD2&quot;, &quot;DYRK1A&quot;, &quot;PTEN&quot;, &quot;SETD5&quot;)
complexplot_perturbation_factor(gibbs_PM$Gamma_pm[-nrow(gibbs_PM$Gamma_pm), ],
                                gibbs_PM$beta_pm[-nrow(gibbs_PM$beta_pm), ],
                                marker_names = KO_names,
                                reorder_markers = targets,
                                reorder_factors = c(4, 9, 16))</code></pre>
<p><img src="gsfa_result_interpret_LUHMES_files/figure-html/Fig_5A-1.png" width="384" style="display: block; margin: auto;" /></p>
<pre class="r"><code>## Similar visualization using GSFA built-in functions:
GSFA::dotplot_beta_PIP(fit,
                       target_names = KO_names,
                       reorder_targets = targets, reorder_factors = c(4, 9, 16))</code></pre>
</div>
<div id="factor-perturbation-association-p-values" class="section level2">
<h2><span class="header-section-number">2.2</span> Factor-perturbation association p values</h2>
<p>We can also assess the correlations between each pair of perturbation and inferred factor.<br />
The distribution of correlation p values show significant signals.</p>
<pre class="r"><code>gibbs_res_tb &lt;- make_gibbs_res_tb(gibbs_PM, G_mat, compute_pve = F)
heatmap_matrix &lt;- gibbs_res_tb %&gt;% select(starts_with(&quot;pval&quot;))
rownames(heatmap_matrix) &lt;- 1:nrow(heatmap_matrix)
colnames(heatmap_matrix) &lt;- colnames(G_mat)

summ_pvalues(unlist(heatmap_matrix),
             title_text = &quot;GSFA\n(15 Targets x 20 Factors)&quot;)</code></pre>
<p><img src="gsfa_result_interpret_LUHMES_files/figure-html/unnamed-chunk-7-1.png" width="960" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="factor-interpretation" class="section level1">
<h1><span class="header-section-number">3</span> Factor Interpretation</h1>
<div id="correlation-within-factors" class="section level2">
<h2><span class="header-section-number">3.1</span> Correlation within factors</h2>
<p>Since the GSFA model does not enforce orthogonality among factors, we first inspect the pairwise correlation within them to see if there is any redundancy. As we can see below, the inferred factors are mostly independent of each other.</p>
<pre class="r"><code>plot_pairwise.corr_heatmap(input_mat_1 = gibbs_PM$Z_pm,
                           corr_type = &quot;pearson&quot;,
                           name_1 = &quot;Pairwise correlation within factors (Z)&quot;,
                           label_size = 10)</code></pre>
<p><img src="gsfa_result_interpret_LUHMES_files/figure-html/unnamed-chunk-8-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>plot_pairwise.corr_heatmap(input_mat_1 = (gibbs_PM$F_pm &gt; 0.95) * 1,
                           corr_type = &quot;jaccard&quot;,
                           name_1 = &quot;Pairwise correlation within \nbinarized gene loadings (F_pm &gt; 0.95)&quot;,
                           label_size = 10)</code></pre>
<p><img src="gsfa_result_interpret_LUHMES_files/figure-html/unnamed-chunk-9-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="gene-loading-in-factors" class="section level2">
<h2><span class="header-section-number">3.2</span> Gene loading in factors</h2>
<p>To understand these latent factors, we inspect the loadings (weights) of several marker genes for neuron maturation and differentiation in them.</p>
<div style="border: 1px solid #ddd; padding: 0px; overflow-y: scroll; height:400px; overflow-x: scroll; width:100%; ">
<table class="table" style="margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
gene_name
</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
type
</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
protein_name
</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
gene_ID
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
TP53
</td>
<td style="text-align:left;">
Cell proliferation
</td>
<td style="text-align:left;">
Tumor protein p53
</td>
<td style="text-align:left;">
ENSG00000141510
</td>
</tr>
<tr>
<td style="text-align:left;">
CDK4
</td>
<td style="text-align:left;">
Cell proliferation
</td>
<td style="text-align:left;">
Cyclin dependent kinase 4
</td>
<td style="text-align:left;">
ENSG00000135446
</td>
</tr>
<tr>
<td style="text-align:left;">
NES
</td>
<td style="text-align:left;">
Neural progenitor cell
</td>
<td style="text-align:left;">
Nestin
</td>
<td style="text-align:left;">
ENSG00000132688
</td>
</tr>
<tr>
<td style="text-align:left;">
STMN2
</td>
<td style="text-align:left;">
Mature neuron
</td>
<td style="text-align:left;">
Stathmin-2
</td>
<td style="text-align:left;">
ENSG00000104435
</td>
</tr>
<tr>
<td style="text-align:left;">
MAP2
</td>
<td style="text-align:left;">
Mature neuron
</td>
<td style="text-align:left;">
Microtubule associated protein 2
</td>
<td style="text-align:left;">
ENSG00000078018
</td>
</tr>
<tr>
<td style="text-align:left;">
DPYSL3
</td>
<td style="text-align:left;">
Mature neuron
</td>
<td style="text-align:left;">
Dihydropyrimidinase like 3
</td>
<td style="text-align:left;">
ENSG00000113657
</td>
</tr>
<tr>
<td style="text-align:left;">
MAP1B
</td>
<td style="text-align:left;">
Mature neuron
</td>
<td style="text-align:left;">
Microtubule associated protein 1B
</td>
<td style="text-align:left;">
ENSG00000131711
</td>
</tr>
<tr>
<td style="text-align:left;">
CRABP2
</td>
<td style="text-align:left;">
Mature neuron
</td>
<td style="text-align:left;">
Cellular retinoic acid binding protein 2
</td>
<td style="text-align:left;">
ENSG00000143320
</td>
</tr>
<tr>
<td style="text-align:left;">
NEFL
</td>
<td style="text-align:left;">
Mature neuron
</td>
<td style="text-align:left;">
Neurofilament Light Chain
</td>
<td style="text-align:left;">
ENSG00000277586
</td>
</tr>
<tr>
<td style="text-align:left;">
ZEB2
</td>
<td style="text-align:left;">
Mature neuron
</td>
<td style="text-align:left;">
Zinc finger E-box binding homeobox 2
</td>
<td style="text-align:left;">
ENSG00000169554
</td>
</tr>
<tr>
<td style="text-align:left;">
ITM2C
</td>
<td style="text-align:left;">
Negative regulation of neuron projection
</td>
<td style="text-align:left;">
Integral membrane protein 2C
</td>
<td style="text-align:left;">
ENSG00000135916
</td>
</tr>
<tr>
<td style="text-align:left;">
CNTN2
</td>
<td style="text-align:left;">
Negative regulation of neuron projection
</td>
<td style="text-align:left;">
Contactin-2
</td>
<td style="text-align:left;">
ENSG00000184144
</td>
</tr>
<tr>
<td style="text-align:left;">
DRAXIN
</td>
<td style="text-align:left;">
Negative regulation of neuron projection
</td>
<td style="text-align:left;">
Dorsal inhibitory axon guidance protein
</td>
<td style="text-align:left;">
ENSG00000162490
</td>
</tr>
<tr>
<td style="text-align:left;">
HDAC2
</td>
<td style="text-align:left;">
Negative regulation of neuron projection
</td>
<td style="text-align:left;">
Histone deacetylase 2
</td>
<td style="text-align:left;">
ENSG00000196591
</td>
</tr>
</tbody>
</table>
</div>
<p>We visualize both the gene PIPs (dot size) and gene weights (dot color) in all factors (Figure S6B):</p>
<pre class="r"><code>complexplot_gene_factor(genes_df, interest_df, gibbs_PM$F_pm, gibbs_PM$W_pm)</code></pre>
<p><img src="gsfa_result_interpret_LUHMES_files/figure-html/Fig_S6B-1.png" width="960" style="display: block; margin: auto;" /></p>
<p>A closer look at some factors that are associated with perturbations (Figure 5C):</p>
<pre class="r"><code>complexplot_gene_factor(genes_df, interest_df, gibbs_PM$F_pm, gibbs_PM$W_pm,
                        reorder_factors = c(4, 9, 16))</code></pre>
<p><img src="gsfa_result_interpret_LUHMES_files/figure-html/Fig_5C-1.png" width="624" style="display: block; margin: auto;" /></p>
</div>
<div id="go-enrichment-analysis-in-factors" class="section level2">
<h2><span class="header-section-number">3.3</span> GO enrichment analysis in factors</h2>
<p>To further characterize these latent factors, we perform GO (gene ontology) enrichment analysis of genes loaded on the factors using <code>WebGestalt</code>.</p>
<p>Foreground genes: genes w/ non-zero loadings in each factor (gene PIP &gt; 0.95);<br />
Background genes: all 6000 genes used in GSFA;<br />
Statistical test: hypergeometric test (over-representation test);<br />
Gene sets: GO Slim &quot;Biological Process&quot; (non-redundant).</p>
<pre class="r"><code>## The &quot;WebGestaltR&quot; tool needs Internet connection.
enrich_db &lt;- &quot;geneontology_Biological_Process_noRedundant&quot;
PIP_mat &lt;- gibbs_PM$F_pm
enrich_res_by_factor &lt;- list()
for (i in 1:ncol(PIP_mat)){
  enrich_res_by_factor[[i]] &lt;- 
    WebGestaltR::WebGestaltR(enrichMethod = &quot;ORA&quot;,
                             organism = &quot;hsapiens&quot;,
                             enrichDatabase = enrich_db,
                             interestGene = genes_df[PIP_mat[, i] &gt; 0.05, ]$ID,
                             interestGeneType = &quot;ensembl_gene_id&quot;,
                             referenceGene = genes_df$ID,
                             referenceGeneType = &quot;ensembl_gene_id&quot;,
                             isOutput = F)
}</code></pre>
<p>Several GO “biological process” terms related to neuronal development are enriched in factors 4, 9, and 16 (Figure 5D):</p>
<pre class="r"><code>factor_indx &lt;- 4
terms_of_interest &lt;- c(&quot;regulation of ion transmembrane transport&quot;,
                       &quot;regulation of trans-synaptic signaling&quot;,
                       &quot;axon development&quot;,
                       &quot;regulation of neuron projection development&quot;)
barplot_top_enrich_terms(enrich_res_by_factor[[factor_indx]],
                         terms_of_interest = terms_of_interest,
                         str_wrap_length = 35, pval_max = 8, FC_max = 6) +
  labs(title = paste0(&quot;Factor &quot;, factor_indx),
       x = &quot;Fold of enrichment&quot;)</code></pre>
<p><img src="gsfa_result_interpret_LUHMES_files/figure-html/Fig_5D_factor4-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>factor_indx &lt;- 9
terms_of_interest &lt;- c(&quot;actin filament organization&quot;,
                       &quot;cell fate commitment&quot;,
                       &quot;regulation of neuron projection development&quot;,
                       &quot;regulation of cell morphogenesis&quot;)
barplot_top_enrich_terms(enrich_res_by_factor[[factor_indx]],
                         terms_of_interest = terms_of_interest,
                         str_wrap_length = 35, pval_max = 8, FC_max = 6) +
  labs(title = paste0(&quot;Factor &quot;, factor_indx),
       x = &quot;Fold of enrichment&quot;)</code></pre>
<p><img src="gsfa_result_interpret_LUHMES_files/figure-html/Fig_5D_factor9-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>factor_indx &lt;- 16
terms_of_interest &lt;- c(&quot;developmental growth involved in morphogenesis&quot;,
                       &quot;axon development&quot;)
barplot_top_enrich_terms(enrich_res_by_factor[[factor_indx]],
                         terms_of_interest = terms_of_interest, 
                         str_wrap_length = 35, pval_max = 8, FC_max = 6) +
  labs(title = paste0(&quot;Factor &quot;, factor_indx),
       x = &quot;Fold of enrichment&quot;)</code></pre>
<p><img src="gsfa_result_interpret_LUHMES_files/figure-html/Fig_5D_factor16-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="deg-interpretation" class="section level1">
<h1><span class="header-section-number">4</span> DEG Interpretation</h1>
<p>In GSFA, differential expression analysis can be performed based on the LFSR method. Here we evaluate the specific downstream genes affected by the perturbations detected by GSFA.</p>
<p>We also performed several other differential expression methods for comparison, including scMAGeCK-LR, MAST, and DESeq.</p>
<div id="number-of-degs-detected-by-different-methods" class="section level2">
<h2><span class="header-section-number">4.1</span> Number of DEGs detected by different methods</h2>
<pre class="r"><code>fdr_cutoff &lt;- 0.05
lfsr_cutoff &lt;- 0.05</code></pre>
<div style="border: 1px solid #ddd; padding: 5px; overflow-x: scroll; width:100%; ">
<table class="table" style="margin-left: auto; margin-right: auto;">
<caption>
Number of DEGs detected by GSFA under each perturbation:
</caption>
<tbody>
<tr>
<td style="text-align:left;">
KO
</td>
<td style="text-align:left;">
ADNP
</td>
<td style="text-align:left;">
ARID1B
</td>
<td style="text-align:left;">
ASH1L
</td>
<td style="text-align:left;">
CHD2
</td>
<td style="text-align:left;">
CHD8
</td>
</tr>
<tr>
<td style="text-align:left;">
Num_genes
</td>
<td style="text-align:left;">
790
</td>
<td style="text-align:left;">
345
</td>
<td style="text-align:left;">
341
</td>
<td style="text-align:left;">
773
</td>
<td style="text-align:left;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
KO
</td>
<td style="text-align:left;">
CTNND2
</td>
<td style="text-align:left;">
DYRK1A
</td>
<td style="text-align:left;">
HDAC5
</td>
<td style="text-align:left;">
MECP2
</td>
<td style="text-align:left;">
MYT1L
</td>
</tr>
<tr>
<td style="text-align:left;">
Num_genes
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
111
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
KO
</td>
<td style="text-align:left;">
Nontargeting
</td>
<td style="text-align:left;">
POGZ
</td>
<td style="text-align:left;">
PTEN
</td>
<td style="text-align:left;">
RELN
</td>
<td style="text-align:left;">
SETD5
</td>
</tr>
<tr>
<td style="text-align:left;">
Num_genes
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
898
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
389
</td>
</tr>
</tbody>
</table>
</div>
<pre class="r"><code>deseq_list &lt;- readRDS(paste0(data_folder, &quot;DE_result_DESeq2.rds&quot;))
deseq_signif_counts &lt;- sapply(deseq_list, function(x){filter(x, FDR &lt; fdr_cutoff) %&gt;% nrow()})</code></pre>
<pre class="r"><code>mast_list &lt;- readRDS(paste0(data_folder, &quot;DE_result_MAST.rds&quot;))
mast_signif_counts &lt;- sapply(mast_list, function(x){filter(x, FDR &lt; fdr_cutoff) %&gt;% nrow()})</code></pre>
<pre class="r"><code>scmageck_res &lt;- readRDS(paste0(data_folder, &quot;DE_result_scMAGeCK.rds&quot;))
colnames(scmageck_res$fdr)[colnames(scmageck_res$fdr) == &quot;NegCtrl&quot;] &lt;- &quot;Nontargeting&quot;
scmageck_signif_counts &lt;- colSums(scmageck_res$fdr[, KO_names] &lt; fdr_cutoff)</code></pre>
<pre class="r"><code>dge_comparison_df &lt;- data.frame(Perturbation = names(lfsr_signif_num),
                                GSFA = lfsr_signif_num,
                                scMAGeCK = scmageck_signif_counts,
                                DESeq2 = deseq_signif_counts,
                                MAST = mast_signif_counts)
dge_comparison_df$Perturbation[dge_comparison_df$Perturbation == &quot;Nontargeting&quot;] &lt;- &quot;NegCtrl&quot;</code></pre>
<p>Number of DEGs detected under each perturbation using 4 different methods (Figure 5E):</p>
<pre class="r"><code>dge_plot_df &lt;- reshape2::melt(dge_comparison_df, id.var = &quot;Perturbation&quot;,
                              variable.name = &quot;Method&quot;, value.name = &quot;Num_DEGs&quot;)
dge_plot_df$Perturbation &lt;- factor(dge_plot_df$Perturbation,
                                   levels = c(&quot;NegCtrl&quot;, KO_names[KO_names!=&quot;Nontargeting&quot;]))
ggplot(dge_plot_df, aes(x = Perturbation, y = Num_DEGs+1, fill = Method)) +
  geom_bar(position = &quot;dodge&quot;, stat = &quot;identity&quot;) +
  geom_text(aes(label = Num_DEGs), position=position_dodge(width=0.9), vjust=-0.25) +
  scale_y_log10() +
  scale_fill_brewer(palette = &quot;Set2&quot;) +
  labs(x = &quot;Target gene&quot;,
       y = &quot;Number of DEGs&quot;,
       title = &quot;Number of DEGs detected by different methods&quot;) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 12),
        legend.position = &quot;bottom&quot;,
        legend.text = element_text(size = 13))</code></pre>
<p><img src="gsfa_result_interpret_LUHMES_files/figure-html/Fig_5E-1.png" width="1248" style="display: block; margin: auto;" /></p>
<p>Compared with other differential expression analysis methods, GSFA detected the most DEGs for 5 out of 7 gene targets that have significant effects.</p>
</div>
<div id="perturbation-effects-on-marker-genes" class="section level2">
<h2><span class="header-section-number">4.2</span> Perturbation effects on marker genes</h2>
<p>To better understand the functions of these 7 target genes, we examined their effects on marker genes for neuron maturation and differentiation.</p>
<div id="gsfa" class="section level3">
<h3><span class="header-section-number">4.2.1</span> GSFA</h3>
<p>Here are the summarized effects of perturbations on marker genes estimated by GSFA (Figure 5G).</p>
<p>As we can see, knockdown of ADNP, ASH1L, CHD2, and DYRK1A has mostly negative effects on mature neuronal markers, and positive effects on negative regulators of neuron projection, indicating delayed neuron maturation.</p>
<p>Knockdown of PTEN and SETD5 has the opposite pattern, which indicates accelerated neuron maturation.</p>
<pre class="r"><code>targets &lt;- c(&quot;ADNP&quot;, &quot;ARID1B&quot;, &quot;ASH1L&quot;, &quot;CHD2&quot;, &quot;DYRK1A&quot;, &quot;PTEN&quot;, &quot;SETD5&quot;)
complexplot_gene_perturbation(genes_df, interest_df,
                              targets = targets,
                              lfsr_mat = lfsr_mat,
                              effect_mat = gibbs_PM$W_pm %*%
                                t(gibbs_PM$beta_pm[-nrow(gibbs_PM$beta_pm), ]))</code></pre>
<p><img src="gsfa_result_interpret_LUHMES_files/figure-html/Fig_5G-1.png" width="624" style="display: block; margin: auto;" /></p>
<pre class="r"><code>## Similar visualization using GSFA built-in functions:
GSFA::dotplot_total_effect(fit,
                           gene_indices = match(interest_df$gene_ID, rownames(lfsr_mat)),
                           gene_names = interest_df$gene_name,
                           reorder_targets = targets, 
                           plot_max_score = 0.2)</code></pre>
</div>
<div id="scmageck" class="section level3">
<h3><span class="header-section-number">4.2.2</span> scMAGeCK</h3>
<p>Here are scMAGeCK estimated effects of perturbations on marker genes (Figure 5H):</p>
<pre class="r"><code>score_mat &lt;- scmageck_res$score
fdr_mat &lt;- scmageck_res$fdr
complexplot_gene_perturbation(genes_df, interest_df,
                              targets = targets,
                              lfsr_mat = fdr_mat, lfsr_name = &quot;FDR&quot;,
                              effect_mat = score_mat, effect_name = &quot;scMAGeCK\nselection score&quot;, 
                              score_break = c(-0.2, 0, 0.2),
                              color_break = c(&quot;blue3&quot;, &quot;grey90&quot;, &quot;red3&quot;))</code></pre>
<p><img src="gsfa_result_interpret_LUHMES_files/figure-html/Fig_5H-1.png" width="624" style="display: block; margin: auto;" /></p>
</div>
<div id="deseq2" class="section level3">
<h3><span class="header-section-number">4.2.3</span> DESeq2</h3>
<pre class="r"><code>FC_mat &lt;- matrix(nrow = nrow(interest_df), ncol = length(targets))
rownames(FC_mat) &lt;- interest_df$gene_name
colnames(FC_mat) &lt;- targets
fdr_mat &lt;- FC_mat
for (m in targets){
  FC_mat[, m] &lt;- deseq_list[[m]]$log2FoldChange[match(interest_df$gene_ID,
                                                      deseq_list[[m]]$geneID)]
  fdr_mat[, m] &lt;- deseq_list[[m]]$FDR[match(interest_df$gene_ID, deseq_list[[m]]$geneID)]
}</code></pre>
<p>Here are DESeq2 estimated effects of perturbations on marker genes (Figure S5C):</p>
<pre class="r"><code>complexplot_gene_perturbation(genes_df, interest_df,
                              targets = targets,
                              lfsr_mat = fdr_mat, lfsr_name = &quot;FDR&quot;,
                              effect_mat = FC_mat, effect_name = &quot;DESeq2 log2FC&quot;,
                              score_break = c(-0.4, 0, 0.4),
                              color_break = c(&quot;blue3&quot;, &quot;grey90&quot;, &quot;red3&quot;))</code></pre>
<p><img src="gsfa_result_interpret_LUHMES_files/figure-html/Fig_S5C-1.png" width="624" style="display: block; margin: auto;" /></p>
</div>
<div id="mast" class="section level3">
<h3><span class="header-section-number">4.2.4</span> MAST</h3>
<pre class="r"><code>FC_mat &lt;- matrix(nrow = nrow(interest_df), ncol = length(targets))
rownames(FC_mat) &lt;- interest_df$gene_name
colnames(FC_mat) &lt;- targets
fdr_mat &lt;- FC_mat
for (m in targets){
  FC_mat[, m] &lt;- mast_list[[m]]$logFC[match(interest_df$gene_ID, mast_list[[m]]$geneID)]
  fdr_mat[, m] &lt;- mast_list[[m]]$FDR[match(interest_df$gene_ID, mast_list[[m]]$geneID)]
}</code></pre>
<p>MAST estimated effects of perturbations on marker genes (Figure S5D):</p>
<pre class="r"><code>complexplot_gene_perturbation(genes_df, interest_df,
                              targets = targets,
                              lfsr_mat = fdr_mat, lfsr_name = &quot;FDR&quot;,
                              effect_mat = FC_mat, effect_name = &quot;MAST logFC&quot;,
                              score_break = c(-0.4, 0, 0.4),
                              color_break = c(&quot;blue3&quot;, &quot;grey90&quot;, &quot;red3&quot;))</code></pre>
<p><img src="gsfa_result_interpret_LUHMES_files/figure-html/Fig_S5D-1.png" width="624" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="go-enrichment-in-degs" class="section level2">
<h2><span class="header-section-number">4.3</span> GO enrichment in DEGs</h2>
<p>We further examine these DEGs for enrichment of relevant biological processes through GO enrichment analysis.</p>
<p>Foreground genes: Genes w/ GSFA LFSR &lt; 0.05 under each perturbation;<br />
Background genes: all 6000 genes used in GSFA;<br />
Statistical test: hypergeometric test (over-representation test);<br />
Gene sets: Gene ontology &quot;Biological Process&quot; (non-redundant).</p>
<pre class="r"><code>## The &quot;WebGestaltR&quot; tool needs Internet connection.
targets &lt;- names(lfsr_signif_num)[lfsr_signif_num &gt; 0]
enrich_db &lt;- &quot;geneontology_Biological_Process_noRedundant&quot;
enrich_res &lt;- list()
for (i in targets){
  print(i)
  interest_genes &lt;- genes_df %&gt;% mutate(lfsr = lfsr_mat[, i]) %&gt;%
    filter(lfsr &lt; lfsr_cutoff) %&gt;% pull(ID)
  enrich_res[[i]] &lt;- 
    WebGestaltR::WebGestaltR(enrichMethod = &quot;ORA&quot;,
                             organism = &quot;hsapiens&quot;,
                             enrichDatabase = enrich_db,
                             interestGene = interest_genes,
                             interestGeneType = &quot;ensembl_gene_id&quot;,
                             referenceGene = genes_df$ID,
                             referenceGeneType = &quot;ensembl_gene_id&quot;,
                             isOutput = F)
}</code></pre>
<pre class="r"><code>signif_GO_list &lt;- list()
for (i in names(enrich_res)) {
  signif_GO_list[[i]] &lt;- enrich_res[[i]] %&gt;%
    dplyr::filter(FDR &lt; 0.05) %&gt;%
    dplyr::select(geneSet, description, size, enrichmentRatio, pValue) %&gt;%
    mutate(target = i)
}
signif_term_df &lt;- do.call(rbind, signif_GO_list) %&gt;%
  group_by(geneSet, description, size) %&gt;%
  summarise(pValue = min(pValue)) %&gt;%
  ungroup()

abs_FC_colormap &lt;- circlize::colorRamp2(breaks = c(0, 3, 6),
                                        colors = c(&quot;grey95&quot;, &quot;#77d183&quot;, &quot;#255566&quot;))</code></pre>
<pre class="r"><code>targets &lt;- names(enrich_res)
enrich_table &lt;- data.frame(matrix(nrow = nrow(signif_term_df),
                                  ncol = length(targets)),
                           row.names = signif_term_df$geneSet)
colnames(enrich_table) &lt;- targets
for (i in 1:ncol(enrich_table)){
  m &lt;- colnames(enrich_table)[i]
  enrich_df &lt;- enrich_res[[m]] %&gt;% dplyr::filter(FDR &lt; 0.05)
  enrich_table[enrich_df$geneSet, i] &lt;- enrich_df$enrichmentRatio
}
rownames(enrich_table) &lt;- signif_term_df$description</code></pre>
<p>Here are selected GO &quot;biological process&quot;&quot; terms and their folds of enrichment in DEGs detected by GSFA (Figure S5E):<br />
(In the code below, we omitted the content in <code>terms_of_interest_df</code> as one can subset the <code>enrich_table</code> with any terms of their choice.)</p>
<pre class="r"><code>interest_enrich_table &lt;- enrich_table[terms_of_interest_df$description,
                                      colnames(enrich_table) != &quot;DYRK1A&quot;]
interest_enrich_table[is.na(interest_enrich_table)] &lt;- 0

map &lt;- Heatmap(abs(interest_enrich_table),
               name = &quot;Fold of enrichment&quot;,
               col = abs_FC_colormap,
               na_col = &quot;grey90&quot;,
               row_title = NULL, column_title = NULL,
               cluster_rows = F, cluster_columns = F,
               show_row_dend = F, show_column_dend = F,
               show_heatmap_legend = T,
               row_names_gp = gpar(fontsize = 10.5),
               column_names_rot = 45,
               column_names_side = &quot;top&quot;,
               width = unit(6, &quot;cm&quot;))
draw(map, heatmap_legend_side = &quot;bottom&quot;)</code></pre>
<p><img src="gsfa_result_interpret_LUHMES_files/figure-html/Fig_S5E-1.png" width="768" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="session-information" class="section level1">
<h1><span class="header-section-number">5</span> Session Information</h1>
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.0.4 (2021-02-15)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Scientific Linux 7.4 (Nitrogen)

Matrix products: default
BLAS/LAPACK: /software/openblas-0.3.13-el7-x86_64/lib/libopenblas_haswellp-r0.3.13.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] grid      stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] lattice_0.20-41      WebGestaltR_0.4.4    kableExtra_1.3.4    
 [4] ComplexHeatmap_2.6.2 gridExtra_2.3        forcats_0.5.1       
 [7] stringr_1.4.0        dplyr_1.0.4          purrr_0.3.4         
[10] readr_1.4.0          tidyr_1.1.2          tibble_3.0.6        
[13] ggplot2_3.3.3        tidyverse_1.3.0      Matrix_1.3-2        
[16] data.table_1.14.0   

loaded via a namespace (and not attached):
 [1] matrixStats_0.58.0  fs_1.5.0            lubridate_1.7.9.2  
 [4] doParallel_1.0.16   webshot_0.5.2       RColorBrewer_1.1-2 
 [7] httr_1.4.2          doRNG_1.8.2         tools_4.0.4        
[10] backports_1.2.1     bslib_0.2.4         utf8_1.1.4         
[13] R6_2.5.0            DBI_1.1.1           BiocGenerics_0.36.1
[16] colorspace_2.0-0    GetoptLong_1.0.5    withr_2.4.1        
[19] tidyselect_1.1.0    compiler_4.0.4      cli_2.3.1          
[22] rvest_0.3.6         Cairo_1.5-12.2      xml2_1.3.2         
[25] labeling_0.4.2      sass_0.3.1          scales_1.1.1       
[28] apcluster_1.4.8     systemfonts_1.0.1   digest_0.6.27      
[31] rmarkdown_2.11      svglite_2.0.0       pkgconfig_2.0.3    
[34] htmltools_0.5.1.1   highr_0.8           dbplyr_2.1.0       
[37] rlang_0.4.10        GlobalOptions_0.1.2 readxl_1.3.1       
[40] rstudioapi_0.13     farver_2.0.3        shape_1.4.5        
[43] jquerylib_0.1.3     generics_0.1.0      jsonlite_1.7.2     
[46] magrittr_2.0.1      Rcpp_1.0.6          munsell_0.5.0      
[49] S4Vectors_0.28.1    fansi_0.4.2         lifecycle_1.0.0    
[52] whisker_0.4         stringi_1.5.3       yaml_2.2.1         
[55] plyr_1.8.6          parallel_4.0.4      crayon_1.4.1       
[58] haven_2.3.1         pander_0.6.4        circlize_0.4.12    
[61] hms_1.0.0           knitr_1.34          pillar_1.5.0       
[64] igraph_1.2.6        rjson_0.2.20        rngtools_1.5       
[67] reshape2_1.4.4      codetools_0.2-18    stats4_4.0.4       
[70] reprex_1.0.0        glue_1.4.2          evaluate_0.14      
[73] modelr_0.1.8        foreach_1.5.1       png_0.1-7          
[76] vctrs_0.3.6         cellranger_1.1.0    gtable_0.3.0       
[79] clue_0.3-58         assertthat_0.2.1    xfun_0.26          
[82] broom_0.7.5         viridisLite_0.3.0   iterators_1.0.13   
[85] IRanges_2.24.1      cluster_2.1.0       ellipsis_0.3.1     </code></pre>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
