<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Yifan Zhou (zhouyf@uchicago.edu)" />


<title>Interpretation of GSFA Results on CD8+ T Cell CROP-seq Data</title>

<script src="site_libs/jquery-3.5.1/jquery-3.5.1.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<script src="site_libs/kePrint-0.0.1/kePrint.js"></script>
<link href="site_libs/lightable-0.0.1/lightable.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">GSFA</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Interpretation of GSFA Results on CD8+ T Cell CROP-seq Data</h1>
<h3 class="subtitle">-- for manuscript</h3>
<h4 class="author">Yifan Zhou (<a href="mailto:zhouyf@uchicago.edu">zhouyf@uchicago.edu</a>)</h4>

</div>


<div id="introduction" class="section level1">
<h1><span class="header-section-number">1</span> Introduction</h1>
<p>This page demonstrates how to visualize and interpret the results from a GSFA run.</p>
<p>We have described in this <a href="https://gradonion.github.io/GSFA_paper/preprocess_and_gsfa_TCells.html">page</a> how to run GSFA on CD8+ T cell CROP-seq data from Shifrut et al.</p>
<p>To recapitulate, the processed dataset consists of 10677 unstimulated T cells and 14278 stimulated T cells.<br />
They were subject to belong to one of the 21 perturbation conditions (CRISPR knock-out of 20 regulators of T cell proliferation or immune checkpoint genes, and negative control).<br />
Top 6000 genes ranked by deviance statistics were kept. A modified two-group GSFA was performed on the data with 20 factors specified, and perturbation effects estimated separately for cells with/without TCR stimulation.</p>
<div id="load-necessary-packages-and-data" class="section level2">
<h2><span class="header-section-number">1.1</span> Load necessary packages and data</h2>
<pre class="r"><code>library(data.table)
library(Matrix)
library(tidyverse)
library(ggplot2)
theme_set(theme_bw() + theme(plot.title = element_text(size = 14, hjust = 0.5),
                             axis.title = element_text(size = 14),
                             axis.text = element_text(size = 12),
                             legend.title = element_text(size = 13),
                             legend.text = element_text(size = 12),
                             panel.grid.minor = element_blank())
)
library(gridExtra)
library(ComplexHeatmap)
library(kableExtra)
library(WebGestaltR)

source(&quot;../R/plotting_functions.R&quot;)
data_folder &lt;- &quot;../data/TCells/&quot;</code></pre>
<p>The first thing we need is the output of GSFA <code>fit_gsfa_multivar_2groups()</code> run. The lighter version containing just the posterior mean estimates and LFSR of perturbation-gene effects is enough. (See <code>R/run_gsfa_TCells_2groups.R</code> for more GSFA run details.)</p>
<p>The association of <span class="math inline">\(Z\)</span> with <span class="math inline">\(G\)</span> were estimated separately for stimulated and unstimulated cells, generating two effect size matrices, <span class="math inline">\(\beta_1\)</span> and <span class="math inline">\(\beta_0\)</span>, and two <span class="math inline">\(LFSR\)</span> matrices that summarized the effects of perturbations on genes for stimulated and unstimulated cells, respectively.</p>
<pre class="r"><code>fit &lt;- readRDS(paste0(data_folder, &quot;gsfa_fit.light.rds&quot;))
gibbs_PM &lt;- fit$posterior_means
lfsr_mat1 &lt;- fit$lfsr1[, -ncol(fit$lfsr1)]
lfsr_mat0 &lt;- fit$lfsr0[, -ncol(fit$lfsr0)]</code></pre>
<p>We also need the cell by perturbation matrix which was used as input <span class="math inline">\(G\)</span> for GSFA.<br />
(The row names of which can also inform us which group each cell belongs to.)</p>
<pre class="r"><code>G_mat &lt;- readRDS(paste0(data_folder, &quot;perturbation_matrix.rds&quot;))</code></pre>
<p>Finally, we load the mapping from gene name to ENSEMBL ID for all 6k genes used in GSFA, as well as selected neuronal marker genes. This is specific to this study and analysis.</p>
<pre class="r"><code>genes_df &lt;- readRDS(paste0(data_folder, &quot;top6k_genes.rds&quot;))
interest_df &lt;- readRDS(paste0(data_folder, &quot;selected_tcell_markers.rds&quot;))
KO_names &lt;- colnames(lfsr_mat1)</code></pre>
</div>
</div>
<div id="factor-perturbation-associations" class="section level1">
<h1><span class="header-section-number">2</span> Factor ~ Perturbation Associations</h1>
<div id="perturbation-effects-on-factors-stimulated-cells" class="section level2">
<h2><span class="header-section-number">2.1</span> Perturbation effects on factors (stimulated cells)</h2>
<p>Fisrt of all, we look at the estimated effects of gene perturbations on factors inferred by GSFA.</p>
<p>We found that targeting of 9 genes, ARID1A, CBLB, CD5, CDKN1B, DGKA, LCP2, RASA2, SOCS1, and TCEB2, has significant effects (PIP &gt; 0.95) on at least 1 of the 20 inferred factors.</p>
<p>Estimated effects of perturbations on factors (Figure S4A):</p>
<pre class="r"><code>dotplot_beta_PIP(t(gibbs_PM$Gamma1_pm), t(gibbs_PM$beta1_pm),
                 marker_names = KO_names,
                 reorder_markers = c(KO_names[KO_names!=&quot;NonTarget&quot;], &quot;NonTarget&quot;),
                 inverse_factors = F) +
  coord_flip()</code></pre>
<p><img src="gsfa_result_interpret_TCells_files/figure-html/Fig_S4A-1.png" width="864" style="display: block; margin: auto;" /></p>
<p>Here is a closer look at the estimated effects of selected perturbations on selected factors (Figure 3A):</p>
<pre class="r"><code>targets &lt;- c(&quot;ARID1A&quot;, &quot;LCP2&quot;, &quot;CD5&quot;, &quot;CBLB&quot;, &quot;RASA2&quot;, 
             &quot;DGKA&quot;, &quot;TCEB2&quot;, &quot;SOCS1&quot;, &quot;CDKN1B&quot;)
complexplot_perturbation_factor(gibbs_PM$Gamma1_pm[-nrow(gibbs_PM$Gamma1_pm), ],
                                gibbs_PM$beta1_pm[-nrow(gibbs_PM$beta1_pm), ],
                                marker_names = KO_names, reorder_markers = targets,
                                reorder_factors = c(2, 4, 9, 12))</code></pre>
<p><img src="gsfa_result_interpret_TCells_files/figure-html/Fig_3A-1.png" width="432" style="display: block; margin: auto;" /></p>
<p>We can also assess the correlations between each pair of perturbation and inferred factor.<br />
The distribution of correlation p values show significant signals in stimulated cells.</p>
<pre class="r"><code>## Indices of stimulated cells:
stim_cells &lt;-
  (1:nrow(G_mat))[startsWith(rownames(G_mat), &quot;D1S&quot;) | 
                    startsWith(rownames(G_mat), &quot;D2S&quot;)]
gibbs_res_tb &lt;- make_gibbs_res_tb(gibbs_PM, G_mat, compute_pve = F,
                                  cell_indx = stim_cells)
heatmap_matrix &lt;- gibbs_res_tb %&gt;% select(starts_with(&quot;pval&quot;))
rownames(heatmap_matrix) &lt;- 1:nrow(heatmap_matrix)
colnames(heatmap_matrix) &lt;- colnames(G_mat)

summ_pvalues(unlist(heatmap_matrix),
             title_text = &quot;GSFA Stimulated\n(21 Targets x 20 Factors)&quot;)</code></pre>
<p><img src="gsfa_result_interpret_TCells_files/figure-html/unnamed-chunk-5-1.png" width="960" style="display: block; margin: auto;" /></p>
</div>
<div id="perturbation-effects-on-factors-unstimulated-cells" class="section level2">
<h2><span class="header-section-number">2.2</span> Perturbation effects on factors (unstimulated cells)</h2>
<p>In unstimulated cells, only three pairs of associations were detected at PIP &gt; 0.95, which is unsurprising given the role of these targeted genes in regulating T cell responses (Figure S4B):</p>
<pre class="r"><code>dotplot_beta_PIP(t(gibbs_PM$Gamma0_pm), t(gibbs_PM$beta0_pm),
                 marker_names = KO_names,
                 reorder_markers = c(KO_names[KO_names!=&quot;NonTarget&quot;], &quot;NonTarget&quot;),
                 inverse_factors = F) +
  coord_flip()</code></pre>
<p><img src="gsfa_result_interpret_TCells_files/figure-html/Fig_S4B-1.png" width="864" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="factor-interpretation" class="section level1">
<h1><span class="header-section-number">3</span> Factor Interpretation</h1>
<div id="correlation-within-factors" class="section level2">
<h2><span class="header-section-number">3.1</span> Correlation within factors</h2>
<p>Since the GSFA model does not enforce orthogonality among factors, we first inspect the pairwise correlation within them to see if there is any redundancy. As we can see below, the inferred factors are mostly independent of each other.</p>
<pre class="r"><code>plot_pairwise.corr_heatmap(input_mat_1 = gibbs_PM$Z_pm,
                           corr_type = &quot;pearson&quot;,
                           name_1 = &quot;Pairwise correlation within factors (Z)&quot;,
                           label_size = 10)</code></pre>
<p><img src="gsfa_result_interpret_TCells_files/figure-html/unnamed-chunk-6-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>plot_pairwise.corr_heatmap(input_mat_1 = (gibbs_PM$F_pm &gt; 0.95) * 1,
                           corr_type = &quot;jaccard&quot;,
                           name_1 = &quot;Pairwise correlation within \nbinarized gene loadings (F_pm &gt; 0.95)&quot;,
                           label_size = 10)</code></pre>
<p><img src="gsfa_result_interpret_TCells_files/figure-html/unnamed-chunk-7-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="gene-loading-in-factors" class="section level2">
<h2><span class="header-section-number">3.2</span> Gene loading in factors</h2>
<p>To understand these latent factors, we inspect the loadings (weights) of several marker genes for T cell activation or proliferation states in them.</p>
<div style="border: 1px solid #ddd; padding: 0px; overflow-y: scroll; height:400px; overflow-x: scroll; width:100%; ">
<table class="table" style="margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
gene_name
</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
type
</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
protein_name
</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
gene_ID
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
IL7R
</td>
<td style="text-align:left;">
T cell resting state
</td>
<td style="text-align:left;">
IL-7 receptor
</td>
<td style="text-align:left;">
ENSG00000168685
</td>
</tr>
<tr>
<td style="text-align:left;">
CCR7
</td>
<td style="text-align:left;">
T cell resting state
</td>
<td style="text-align:left;">
C-C motif chemokine receptor 7
</td>
<td style="text-align:left;">
ENSG00000126353
</td>
</tr>
<tr>
<td style="text-align:left;">
GZMB
</td>
<td style="text-align:left;">
T cell activation
</td>
<td style="text-align:left;">
Granzyme B
</td>
<td style="text-align:left;">
ENSG00000100453
</td>
</tr>
<tr>
<td style="text-align:left;">
IFNG
</td>
<td style="text-align:left;">
T cell activation
</td>
<td style="text-align:left;">
Interferon gamma
</td>
<td style="text-align:left;">
ENSG00000111537
</td>
</tr>
<tr>
<td style="text-align:left;">
CD44
</td>
<td style="text-align:left;">
T cell activation
</td>
<td style="text-align:left;">
CD44
</td>
<td style="text-align:left;">
ENSG00000026508
</td>
</tr>
<tr>
<td style="text-align:left;">
IL2RA
</td>
<td style="text-align:left;">
T cell activation
</td>
<td style="text-align:left;">
IL-2 receptor
</td>
<td style="text-align:left;">
ENSG00000134460
</td>
</tr>
<tr>
<td style="text-align:left;">
XCL1
</td>
<td style="text-align:left;">
T cell activation
</td>
<td style="text-align:left;">
X-C motif chemokine ligand 1
</td>
<td style="text-align:left;">
ENSG00000143184
</td>
</tr>
<tr>
<td style="text-align:left;">
TNFRSF18
</td>
<td style="text-align:left;">
T cell activation
</td>
<td style="text-align:left;">
GITR
</td>
<td style="text-align:left;">
ENSG00000186891
</td>
</tr>
<tr>
<td style="text-align:left;">
ITGAL
</td>
<td style="text-align:left;">
T cell activation
</td>
<td style="text-align:left;">
LFA-1
</td>
<td style="text-align:left;">
ENSG00000005844
</td>
</tr>
<tr>
<td style="text-align:left;">
MKI67
</td>
<td style="text-align:left;">
Cell proliferation
</td>
<td style="text-align:left;">
Marker of proliferation Ki-67
</td>
<td style="text-align:left;">
ENSG00000148773
</td>
</tr>
<tr>
<td style="text-align:left;">
TOP2A
</td>
<td style="text-align:left;">
Cell proliferation
</td>
<td style="text-align:left;">
DNA topoisomerase II alpha
</td>
<td style="text-align:left;">
ENSG00000131747
</td>
</tr>
<tr>
<td style="text-align:left;">
CENPF
</td>
<td style="text-align:left;">
Cell proliferation
</td>
<td style="text-align:left;">
Centromere protein F
</td>
<td style="text-align:left;">
ENSG00000117724
</td>
</tr>
</tbody>
</table>
</div>
<p>We visualize both the gene PIPs (dot size) and gene weights (dot color) in all factors (Figure S4C):</p>
<pre class="r"><code>complexplot_gene_factor(genes_df, interest_df, gibbs_PM$F_pm, gibbs_PM$W_pm)</code></pre>
<p><img src="gsfa_result_interpret_TCells_files/figure-html/Fig_S4C-1.png" width="816" style="display: block; margin: auto;" /></p>
<p>A closer look at some factors that are associated with perturbations (Figure 3C):</p>
<pre class="r"><code>complexplot_gene_factor(genes_df, interest_df, gibbs_PM$F_pm, gibbs_PM$W_pm,
                        reorder_factors = c(2, 4, 9, 12))</code></pre>
<p><img src="gsfa_result_interpret_TCells_files/figure-html/Fig_3C-1.png" width="528" style="display: block; margin: auto;" /></p>
</div>
<div id="go-enrichment-analysis-in-factors" class="section level2">
<h2><span class="header-section-number">3.3</span> GO enrichment analysis in factors</h2>
<p>To further characterize these latent factors, we perform GO (gene ontology) enrichment analysis of genes loaded on the factors using <code>WebGestalt</code>.</p>
<p>Foreground genes: Genes w/ non-zero loadings in each factor (gene PIP &gt; 0.95);<br />
Background genes: all 6000 genes used in GSFA;<br />
Statistical test: hypergeometric test (over-representation test);<br />
Gene sets: GO Slim &quot;Biological Process&quot; (non-redundant).</p>
<pre class="r"><code>## The &quot;WebGestaltR&quot; tool needs Internet connection.
enrich_db &lt;- &quot;geneontology_Biological_Process_noRedundant&quot;
PIP_mat &lt;- gibbs_PM$F_pm
enrich_res &lt;- list()
for (i in 1:ncol(PIP_mat)){
  enrich_res[[i]] &lt;- 
    WebGestaltR::WebGestaltR(enrichMethod = &quot;ORA&quot;,
                             organism = &quot;hsapiens&quot;,
                             enrichDatabase = enrich_db,
                             interestGene = genes_df[PIP_mat[, i] &gt; 0.05, ]$ID,
                             interestGeneType = &quot;ensembl_gene_id&quot;,
                             referenceGene = genes_df$ID,
                             referenceGeneType = &quot;ensembl_gene_id&quot;,
                             isOutput = F)
}</code></pre>
<p>Several GO “biological process” terms related to immune responses or cell cycle are enriched in factors 2, 4, 9, and 12 (Figure 3D):</p>
<pre class="r"><code>factor_indx &lt;- 2
terms_of_interest &lt;- c(&quot;kinetochore organization&quot;, &quot;chromosome segregation&quot;,
                       &quot;cell cycle G2/M phase transition&quot;, &quot;cytokinesis&quot;)
barplot_top_enrich_terms(enrich_res_by_factor[[factor_indx]],
                         terms_of_interest = terms_of_interest,
                         str_wrap_length = 50) +
  labs(title = paste0(&quot;Factor &quot;, factor_indx))</code></pre>
<p><img src="gsfa_result_interpret_TCells_files/figure-html/Fig_3D_factor2-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>factor_indx &lt;- 9
terms_of_interest &lt;- c(&quot;microtubule cytoskeleton organization involved in mitosis&quot;,
                       &quot;chromosome segregation&quot;, &quot;cytokinesis&quot;, &quot;cell cycle checkpoint&quot;)
barplot_top_enrich_terms(enrich_res_by_factor[[factor_indx]],
                         terms_of_interest = terms_of_interest,
                         str_wrap_length = 35) +
  labs(title = paste0(&quot;Factor &quot;, factor_indx))</code></pre>
<p><img src="gsfa_result_interpret_TCells_files/figure-html/Fig_3D_factor9-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>factor_indx &lt;- 4
terms_of_interest &lt;- c(&quot;response to chemokine&quot;, &quot;cell killing&quot;, &quot;leukocyte migration&quot;,
                       &quot;response to interferon-gamma&quot;, &quot;cytokine secretion&quot;)
barplot_top_enrich_terms(enrich_res_by_factor[[factor_indx]],
                         terms_of_interest = terms_of_interest,
                         str_wrap_length = 35) +
  labs(title = paste0(&quot;Factor &quot;, factor_indx))</code></pre>
<p><img src="gsfa_result_interpret_TCells_files/figure-html/Fig_3D_factor4-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>factor_indx &lt;- 12
terms_of_interest &lt;- c(&quot;leukocyte cell-cell adhesion&quot;, &quot;extrinsic apoptotic signaling pathway&quot;,
                       &quot;cell killing&quot;, &quot;T cell activation&quot;, &quot;NIK/NF-kappaB signaling&quot;)
barplot_top_enrich_terms(enrich_res_by_factor[[factor_indx]],
                         terms_of_interest = terms_of_interest,
                         str_wrap_length = 35) +
  labs(title = paste0(&quot;Factor &quot;, factor_indx))</code></pre>
<p><img src="gsfa_result_interpret_TCells_files/figure-html/Fig_3D_factor12-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="deg-interpretation" class="section level1">
<h1><span class="header-section-number">4</span> DEG Interpretation</h1>
<p>In GSFA, differential expression analysis can be performed based on the LFSR method. Here we evaluate the specific downstream genes affected by the perturbations detected by GSFA.</p>
<p>We also performed several other differential expression methods for comparison, including scMAGeCK-LR, MAST, and DESeq.</p>
<p><strong>Here, we compared the DEG results within stimulated cells.</strong></p>
<div id="number-of-degs-detected-by-different-methods" class="section level2">
<h2><span class="header-section-number">4.1</span> Number of DEGs detected by different methods</h2>
<pre class="r"><code>fdr_cutoff &lt;- 0.05
lfsr_cutoff &lt;- 0.05</code></pre>
<div style="border: 1px solid #ddd; padding: 5px; overflow-x: scroll; width:100%; ">
<table class="table" style="margin-left: auto; margin-right: auto;">
<caption>
Number of DEGs detected by GSFA:
</caption>
<tbody>
<tr>
<td style="text-align:left;">
KO
</td>
<td style="text-align:left;">
ARID1A
</td>
<td style="text-align:left;">
BTLA
</td>
<td style="text-align:left;">
C10orf54
</td>
<td style="text-align:left;">
CBLB
</td>
<td style="text-align:left;">
CD3D
</td>
<td style="text-align:left;">
CD5
</td>
<td style="text-align:left;">
CDKN1B
</td>
</tr>
<tr>
<td style="text-align:left;">
Num_genes
</td>
<td style="text-align:left;">
448
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
481
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
586
</td>
<td style="text-align:left;">
710
</td>
</tr>
<tr>
<td style="text-align:left;">
KO
</td>
<td style="text-align:left;">
DGKA
</td>
<td style="text-align:left;">
DGKZ
</td>
<td style="text-align:left;">
HAVCR2
</td>
<td style="text-align:left;">
LAG3
</td>
<td style="text-align:left;">
LCP2
</td>
<td style="text-align:left;">
MEF2D
</td>
<td style="text-align:left;">
NonTarget
</td>
</tr>
<tr>
<td style="text-align:left;">
Num_genes
</td>
<td style="text-align:left;">
148
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
586
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
KO
</td>
<td style="text-align:left;">
PDCD1
</td>
<td style="text-align:left;">
RASA2
</td>
<td style="text-align:left;">
SOCS1
</td>
<td style="text-align:left;">
STAT6
</td>
<td style="text-align:left;">
TCEB2
</td>
<td style="text-align:left;">
TMEM222
</td>
<td style="text-align:left;">
TNFRSF9
</td>
</tr>
<tr>
<td style="text-align:left;">
Num_genes
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
271
</td>
<td style="text-align:left;">
511
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
240
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
0
</td>
</tr>
</tbody>
</table>
</div>
<pre class="r"><code>deseq_list &lt;- readRDS(paste0(data_folder, &quot;DE_result_DESeq2_stimulated.rds&quot;))
deseq_signif_counts &lt;- sapply(deseq_list, function(x){filter(x, FDR &lt; fdr_cutoff) %&gt;% nrow()})</code></pre>
<pre class="r"><code>mast_list &lt;- readRDS(paste0(data_folder, &quot;DE_result_MAST_stimulated.rds&quot;))
mast_signif_counts &lt;- sapply(mast_list, function(x){filter(x, FDR &lt; fdr_cutoff) %&gt;% nrow()})</code></pre>
<pre class="r"><code>scmageck_res &lt;- readRDS(paste0(data_folder, &quot;DE_result_scMAGeCK_stimulated.rds&quot;))
colnames(scmageck_res$fdr)[colnames(scmageck_res$fdr) == &quot;NegCtrl&quot;] &lt;- &quot;NonTarget&quot;
scmageck_signif_counts &lt;- colSums(scmageck_res$fdr[, KO_names] &lt; fdr_cutoff)</code></pre>
<pre class="r"><code>dge_comparison_df &lt;- data.frame(Perturbation = KO_names,
                                GSFA = lfsr_signif_num,
                                scMAGeCK = scmageck_signif_counts,
                                DESeq2 = deseq_signif_counts,
                                MAST = mast_signif_counts)
dge_comparison_df$Perturbation[dge_comparison_df$Perturbation == &quot;NonTarget&quot;] &lt;- &quot;NegCtrl&quot;</code></pre>
<p>Number of DEGs detected under each perturbation using 4 different methods (Figure 4A):<br />
Compared with other differential expression analysis methods, GSFA detected the most DEGs for all 9 gene targets that have significant effects.</p>
<pre class="r"><code>dge_plot_df &lt;- reshape2::melt(dge_comparison_df, id.var = &quot;Perturbation&quot;,
                              variable.name = &quot;Method&quot;, value.name = &quot;Num_DEGs&quot;)
dge_plot_df$Perturbation &lt;- factor(dge_plot_df$Perturbation,
                                   levels = c(&quot;NegCtrl&quot;, KO_names[KO_names!=&quot;NonTarget&quot;]))
ggplot(dge_plot_df, aes(x = Perturbation, y = Num_DEGs+1, fill = Method)) +
  geom_bar(position = &quot;dodge&quot;, stat = &quot;identity&quot;) +
  geom_text(aes(label = Num_DEGs), position=position_dodge(width=0.9), vjust=-0.25) +
  scale_y_log10() +
  scale_fill_brewer(palette = &quot;Set2&quot;) +
  labs(x = &quot;Target genes&quot;,
       y = &quot;Number of DEGs&quot;,
       title = &quot;Number of DEGs detected by different methods&quot;) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 12),
        legend.position = &quot;bottom&quot;,
        legend.text = element_text(size = 13))</code></pre>
<p><img src="gsfa_result_interpret_TCells_files/figure-html/Fig_4A-1.png" width="1248" style="display: block; margin: auto;" /></p>
</div>
<div id="perturbation-effects-on-marker-genes" class="section level2">
<h2><span class="header-section-number">4.2</span> Perturbation effects on marker genes</h2>
<p>To better understand the functions of these 7 target genes, we examined their effects on marker genes for T cell activation or proliferation states.</p>
<div id="gsfa" class="section level3">
<h3><span class="header-section-number">4.2.1</span> GSFA</h3>
<p>Here are the summarized effects of perturbations on marker genes estimated by GSFA (Figure 4D).</p>
<p>Cell cycle:<br />
As we can see, knockout of SOCS1 or CDKN1B has positive effects on cell proliferation markers, indicating increased cell proliferation.</p>
<p>T cell activation and immune response:<br />
Knockout of CD5, CBLB, RASA2 or TCEB2 has mostly positve effects on effector markers, indicating T cell activation; knockout of ARID1A has the opposite pattern.</p>
<pre class="r"><code>complexplot_gene_perturbation(genes_df, interest_df,
                              targets = targets,
                              lfsr_mat = lfsr_mat1,
                              effect_mat = gibbs_PM$W_pm %*%
                                t(gibbs_PM$beta1_pm[-nrow(gibbs_PM$beta1_pm), ]))</code></pre>
<p><img src="gsfa_result_interpret_TCells_files/figure-html/Fig_4D-1.png" width="528" style="display: block; margin: auto;" /></p>
</div>
<div id="scmageck" class="section level3">
<h3><span class="header-section-number">4.2.2</span> scMAGeCK</h3>
<p>Here are scMAGeCK estimated effects of perturbations on marker genes (Figure 4E):</p>
<pre class="r"><code>score_mat &lt;- scmageck_res$score
fdr_mat &lt;- scmageck_res$fdr
complexplot_gene_perturbation(genes_df, interest_df,
                              targets = targets,
                              lfsr_mat = fdr_mat, lfsr_name = &quot;FDR&quot;,
                              effect_mat = score_mat, 
                              effect_name = &quot;scMAGeCK\nselection score&quot;,
                              score_break = c(-0.2, 0, 0.2),
                              color_break = c(&quot;blue3&quot;, &quot;grey90&quot;, &quot;red3&quot;))</code></pre>
<p><img src="gsfa_result_interpret_TCells_files/figure-html/Fig_4E-1.png" width="528" style="display: block; margin: auto;" /></p>
</div>
<div id="deseq2" class="section level3">
<h3><span class="header-section-number">4.2.3</span> DESeq2</h3>
<pre class="r"><code>FC_mat &lt;- matrix(nrow = nrow(interest_df), ncol = length(targets))
rownames(FC_mat) &lt;- interest_df$gene_name
colnames(FC_mat) &lt;- targets
fdr_mat &lt;- FC_mat
for (m in targets){
  FC_mat[, m] &lt;- deseq_list[[m]]$log2FoldChange[match(interest_df$gene_ID, 
                                                      deseq_list[[m]]$geneID)]
  fdr_mat[, m] &lt;- deseq_list[[m]]$FDR[match(interest_df$gene_ID, 
                                            deseq_list[[m]]$geneID)]
}</code></pre>
<p>Here are DESeq2 estimated effects of perturbations on marker genes (Figure S4D):</p>
<pre class="r"><code>complexplot_gene_perturbation(genes_df, interest_df,
                              targets = targets,
                              lfsr_mat = fdr_mat, lfsr_name = &quot;FDR&quot;,
                              effect_mat = FC_mat, effect_name = &quot;DESeq2 log2FC&quot;,
                              score_break = c(-0.4, 0, 0.4),
                              color_break = c(&quot;blue3&quot;, &quot;grey90&quot;, &quot;red3&quot;))</code></pre>
<p><img src="gsfa_result_interpret_TCells_files/figure-html/Fig_S4D-1.png" width="528" style="display: block; margin: auto;" /></p>
</div>
<div id="mast" class="section level3">
<h3><span class="header-section-number">4.2.4</span> MAST</h3>
<pre class="r"><code>FC_mat &lt;- matrix(nrow = nrow(interest_df), ncol = length(targets))
rownames(FC_mat) &lt;- interest_df$gene_name
colnames(FC_mat) &lt;- targets
fdr_mat &lt;- FC_mat
for (m in targets){
  FC_mat[, m] &lt;- mast_list[[m]]$logFC[match(interest_df$gene_ID, 
                                            mast_list[[m]]$geneID)]
  fdr_mat[, m] &lt;- mast_list[[m]]$FDR[match(interest_df$gene_ID, 
                                           mast_list[[m]]$geneID)]
}</code></pre>
<p>MAST estimated effects of perturbations on marker genes (Figure S4E):</p>
<pre class="r"><code>complexplot_gene_perturbation(genes_df, interest_df,
                              targets = targets,
                              lfsr_mat = fdr_mat, lfsr_name = &quot;FDR&quot;,
                              effect_mat = FC_mat, effect_name = &quot;MAST logFC&quot;,
                              score_break = c(-0.4, 0, 0.4),
                              color_break = c(&quot;blue3&quot;, &quot;grey90&quot;, &quot;red3&quot;))</code></pre>
<p><img src="gsfa_result_interpret_TCells_files/figure-html/Fig_S4E-1.png" width="528" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="go-enrichment-in-degs" class="section level2">
<h2><span class="header-section-number">4.3</span> GO enrichment in DEGs</h2>
<p>We further examine these DEGs for enrichment of relevant biological processes through GO enrichment analysis.</p>
<p>Foreground genes: Genes w/ GSFA LFSR &lt; 0.05 under each perturbation;<br />
Background genes: all 6000 genes used in GSFA;<br />
Statistical test: hypergeometric test (over-representation test);<br />
Gene sets: GO Slim &quot;Biological Process&quot; (non-redundant).</p>
<pre class="r"><code>## The &quot;WebGestaltR&quot; tool needs Internet connection.
targets &lt;- names(lfsr_signif_num)[lfsr_signif_num &gt; 0]
enrich_db &lt;- &quot;geneontology_Biological_Process_noRedundant&quot;
enrich_res &lt;- list()
for (i in targets){
  print(i)
  interest_genes &lt;- genes_df %&gt;% mutate(lfsr = lfsr_mat1[, i]) %&gt;%
    filter(lfsr &lt; lfsr_cutoff) %&gt;% pull(ID)
  enrich_res[[i]] &lt;- 
    WebGestaltR::WebGestaltR(enrichMethod = &quot;ORA&quot;,
                             organism = &quot;hsapiens&quot;,
                             enrichDatabase = enrich_db,
                             interestGene = interest_genes,
                             interestGeneType = &quot;ensembl_gene_id&quot;,
                             referenceGene = genes_df$ID,
                             referenceGeneType = &quot;ensembl_gene_id&quot;,
                             isOutput = F)
}</code></pre>
<pre class="r"><code>signif_GO_list &lt;- list()
for (i in names(enrich_res)) {
  signif_GO_list[[i]] &lt;- enrich_res[[i]] %&gt;%
    dplyr::filter(FDR &lt; 0.05) %&gt;%
    dplyr::select(geneSet, description, size, enrichmentRatio, pValue) %&gt;%
    mutate(target = i)
}
signif_term_df &lt;- do.call(rbind, signif_GO_list) %&gt;%
  group_by(geneSet, description, size) %&gt;%
  summarise(pValue = min(pValue)) %&gt;%
  ungroup()

abs_FC_colormap &lt;- circlize::colorRamp2(breaks = c(0, 3, 6),
                                        colors = c(&quot;grey95&quot;, &quot;#77d183&quot;, &quot;#255566&quot;))</code></pre>
<pre class="r"><code>enrich_table &lt;- data.frame(matrix(nrow = nrow(signif_term_df),
                                  ncol = length(targets)),
                           row.names = signif_term_df$geneSet)
colnames(enrich_table) &lt;- targets
for (i in 1:ncol(enrich_table)){
  m &lt;- colnames(enrich_table)[i]
  enrich_df &lt;- enrich_res[[m]] %&gt;% filter(enrichmentRatio &gt; 2)
  enrich_table[enrich_df$geneSet, i] &lt;- enrich_df$enrichmentRatio
}
rownames(enrich_table) &lt;- signif_term_df$description</code></pre>
<p>Here are selected GO &quot;biological process&quot;&quot; terms and their folds of enrichment in DEGs detected by GSFA (Figure 4C).<br />
(In the code below, we omitted the content in <code>terms_of_interest_df</code> as one can subset the <code>enrich_table</code> with any terms of their choice.)</p>
<pre class="r"><code>interest_enrich_table &lt;- enrich_table[terms_of_interest_df$description, ]
interest_enrich_table[is.na(interest_enrich_table)] &lt;- 0

map &lt;- Heatmap(interest_enrich_table,
               name = &quot;Fold of enrichment&quot;,
               col = abs_FC_colormap,
               row_title = NULL, column_title = NULL,
               cluster_rows = F, cluster_columns = F,
               show_row_dend = F, show_column_dend = F,
               show_heatmap_legend = T,
               row_names_gp = gpar(fontsize = 10.5),
               column_names_rot = 45,
               width = unit(7, &quot;cm&quot;))
draw(map, heatmap_legend_side = &quot;right&quot;)</code></pre>
<p><img src="gsfa_result_interpret_TCells_files/figure-html/Fig_4C-1.png" width="864" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="session-information" class="section level1">
<h1><span class="header-section-number">5</span> Session Information</h1>
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.0.4 (2021-02-15)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Scientific Linux 7.4 (Nitrogen)

Matrix products: default
BLAS/LAPACK: /software/openblas-0.3.13-el7-x86_64/lib/libopenblas_haswellp-r0.3.13.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] grid      stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] lattice_0.20-41      WebGestaltR_0.4.4    kableExtra_1.3.4    
 [4] ComplexHeatmap_2.6.2 gridExtra_2.3        forcats_0.5.1       
 [7] stringr_1.4.0        dplyr_1.0.4          purrr_0.3.4         
[10] readr_1.4.0          tidyr_1.1.2          tibble_3.0.6        
[13] ggplot2_3.3.3        tidyverse_1.3.0      Matrix_1.3-2        
[16] data.table_1.14.0   

loaded via a namespace (and not attached):
 [1] matrixStats_0.58.0  fs_1.5.0            lubridate_1.7.9.2  
 [4] doParallel_1.0.16   webshot_0.5.2       RColorBrewer_1.1-2 
 [7] httr_1.4.2          doRNG_1.8.2         tools_4.0.4        
[10] backports_1.2.1     bslib_0.2.4         utf8_1.1.4         
[13] R6_2.5.0            DBI_1.1.1           BiocGenerics_0.36.1
[16] colorspace_2.0-0    GetoptLong_1.0.5    withr_2.4.1        
[19] tidyselect_1.1.0    compiler_4.0.4      cli_2.3.1          
[22] rvest_0.3.6         Cairo_1.5-12.2      xml2_1.3.2         
[25] labeling_0.4.2      sass_0.3.1          scales_1.1.1       
[28] apcluster_1.4.8     systemfonts_1.0.1   digest_0.6.27      
[31] rmarkdown_2.11      svglite_2.0.0       pkgconfig_2.0.3    
[34] htmltools_0.5.1.1   highr_0.8           dbplyr_2.1.0       
[37] rlang_0.4.10        GlobalOptions_0.1.2 readxl_1.3.1       
[40] rstudioapi_0.13     farver_2.0.3        shape_1.4.5        
[43] jquerylib_0.1.3     generics_0.1.0      jsonlite_1.7.2     
[46] magrittr_2.0.1      Rcpp_1.0.6          munsell_0.5.0      
[49] S4Vectors_0.28.1    fansi_0.4.2         lifecycle_1.0.0    
[52] whisker_0.4         stringi_1.5.3       yaml_2.2.1         
[55] plyr_1.8.6          parallel_4.0.4      crayon_1.4.1       
[58] haven_2.3.1         pander_0.6.4        circlize_0.4.12    
[61] hms_1.0.0           knitr_1.34          pillar_1.5.0       
[64] igraph_1.2.6        rjson_0.2.20        rngtools_1.5       
[67] reshape2_1.4.4      codetools_0.2-18    stats4_4.0.4       
[70] reprex_1.0.0        glue_1.4.2          evaluate_0.14      
[73] modelr_0.1.8        foreach_1.5.1       png_0.1-7          
[76] vctrs_0.3.6         cellranger_1.1.0    gtable_0.3.0       
[79] clue_0.3-58         assertthat_0.2.1    xfun_0.26          
[82] broom_0.7.5         viridisLite_0.3.0   iterators_1.0.13   
[85] IRanges_2.24.1      cluster_2.1.0       ellipsis_0.3.1     </code></pre>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
