<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Yifan Zhou (zhouyf@uchicago.edu)" />


<title>Application of GSFA on CD8+ T Cell CROP-seq Data</title>

<script src="site_libs/header-attrs-2.20/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<script src="site_libs/kePrint-0.0.1/kePrint.js"></script>
<link href="site_libs/lightable-0.0.1/lightable.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">GSFA</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Application of GSFA on CD8+ T Cell CROP-seq
Data</h1>
<h3 class="subtitle">â€“ for manuscript</h3>
<h4 class="author">Yifan Zhou (<a href="mailto:zhouyf@uchicago.edu"
class="email">zhouyf@uchicago.edu</a>)</h4>

</div>


<div id="introduction" class="section level1" number="1">
<h1><span class="header-section-number">1</span> Introduction</h1>
<p>This page describes the preprocessing of single-cell CRISPR screen
(CROP-seq) dataset in <a
href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6689405/">Shifrut et
al.</a>, and how to run GSFA on the processed data.</p>
<p>Due to the size of data, we recommend running the code in this
tutorial in an R script instead of in R markdown.</p>
<div id="necessary-packages" class="section level2" number="1.1">
<h2><span class="header-section-number">1.1</span> Necessary
packages</h2>
<pre class="r"><code>library(data.table)
library(tidyverse)
library(Matrix)
library(Seurat)
library(GSFA)
library(ggplot2)</code></pre>
</div>
<div id="original-study-and-data-source" class="section level2"
number="1.2">
<h2><span class="header-section-number">1.2</span> Original study and
data source</h2>
<p>Original CROP-seq study:<br />
Genome-wide CRISPR Screens in Primary Human T Cells Reveal Key
Regulators of Immune Function. Cell. (2018).</p>
<p>Data source:<br />
GEO accession: <a
href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE119450">GSE119450</a>,
<code>GSE119450_RAW.tar</code> file.</p>
<p>The study targeted 20 genes, including 12 genes that were identified
to regulate T cell proliferation, and 8 known immune checkpoint genes
with CRISPR knock-out in primary human CD8+ T cells. After CRISPR
targeting, cells either underwent T cell receptor (TCR) stimulation or
not before sequencing. The overall goal is to understand the effects of
these target genes on the transcriptome and on T cell states and
stimulation responses.</p>
</div>
</div>
<div id="preprocessing-of-crop-seq-data" class="section level1"
number="2">
<h1><span class="header-section-number">2</span> Preprocessing of
CROP-seq data</h1>
<div id="merge-experimental-batches" class="section level2"
number="2.1">
<h2><span class="header-section-number">2.1</span> Merge experimental
batches</h2>
<p>The original data came in four batches (donor 1 and 2 <span
class="math inline">\(\times\)</span> unstimulated and stimulated), each
in standard <code>cellranger</code> single-cell RNA-seq output format.
Below is the code to merge all batches of cells together into one
dataset.</p>
<p>Meanwhile, each cell is also assigned its CRISPR perturbation target
based on the gRNA readout. Although each gene was targeted by 2 gRNAs,
and 8 non-targeting gRNAs were designed as negative control, only
gene-level perturbations are assigned to cells, resulting in 21 (20
genes + negative control) perturbation groups in total.<br />
(Each cell contains a unique gRNA, so the assignment is
straight-forward.)</p>
<pre class="r"><code>## Change the following directory to where the downloaded data is:
data_dir &lt;- &quot;/project2/xinhe/yifan/Factor_analysis/Stimulated_T_Cells/GSE119450_RAW/&quot;

filename_tb &lt;- 
  data.frame(experiment = c(&quot;D1S&quot;, &quot;D2S&quot;, &quot;D1N&quot;, &quot;D2N&quot;),
             prefix = c(&quot;GSM3375488_D1S&quot;, &quot;GSM3375490_D2S&quot;, 
                        &quot;GSM3375487_D1N&quot;, &quot;GSM3375489_D2N&quot;),
             stringsAsFactors = F)
seurat_lst &lt;- list()
guide_lst &lt;- list()

for (i in 1:4){
  experiment &lt;- filename_tb$experiment[i]
  prefix &lt;- filename_tb$prefix[i]
  cat(paste0(&quot;Loading data of &quot;, experiment, &quot; ...&quot;))
  cat(&quot;\n\n&quot;)
  feature.names &lt;- data.frame(fread(paste0(data_dir, experiment, &quot;/genes.tsv&quot;),
                                    header = FALSE), stringsAsFactors = FALSE)
  barcode.names &lt;- data.frame(fread(paste0(data_dir, experiment, &quot;/barcodes.tsv&quot;),
                                    header = FALSE), stringsAsFactors = FALSE)
  barcode.names$V2 &lt;- sapply(strsplit(barcode.names$V1, split = &quot;-&quot;),
                             function(x){x[1]})
  # Load the gene count matrix (gene x cell) and annotate the dimension names:
  dm &lt;- readMM(file = paste0(data_dir, experiment, &quot;/matrix.mtx&quot;))
  rownames(dm) &lt;- feature.names$V1
  colnames(dm) &lt;- barcode.names$V2
  
  # Load the meta data of cells:
  metadata &lt;- data.frame(fread(paste0(data_dir, experiment, &quot;/&quot;,
                                      prefix, &quot;_CellBC_sgRNA.csv.gz&quot;),
                               header = T, sep = &#39;,&#39;), check.names = F)
  metadata$gene_target &lt;- sapply(strsplit(metadata$gRNA.ID, split = &quot;[.]&quot;),
                                 function(x){x[3]})
  metadata$guide &lt;- sapply(strsplit(metadata$gRNA.ID, split = &quot;[.]&quot;),
                           function(x){paste0(x[2], &quot;.&quot;, x[3])})

  metadata &lt;- metadata %&gt;% filter(Cell.BC %in% barcode.names$V2)
  targets &lt;- unique(metadata$gene_target)
  targets &lt;- targets[order(targets)]
  
  # Make a cell by perturbation matrix:
  guide_mat &lt;- data.frame(matrix(nrow = nrow(metadata),
                                 ncol = length(targets)))
  rownames(guide_mat) &lt;- metadata$Cell.BC
  colnames(guide_mat) &lt;- targets
  for (m in targets){
    guide_mat[[m]] &lt;- (metadata$gene_target == m) * 1
  }
  guide_lst[[experiment]] &lt;- guide_mat
  
  # Only keep cells with gRNA info:
  dm.cells_w_gRNA &lt;- dm[, metadata$Cell.BC]
  cat(&quot;Dimensions of final gene expression matrix: &quot;)
  cat(dim(dm.cells_w_gRNA))
  cat(&quot;\n\n&quot;)
  
  dm.seurat &lt;- CreateSeuratObject(dm.cells_w_gRNA, project = paste0(&quot;TCells_&quot;, experiment))
  dm.seurat &lt;- AddMetaData(dm.seurat, metadata = guide_mat)
  seurat_lst[[experiment]] &lt;- dm.seurat
}</code></pre>
<p>Loading data of D1S â€¦</p>
<p>Dimensions of final gene expression matrix: 33694 6953</p>
<p>Loading data of D2S â€¦</p>
<p>Dimensions of final gene expression matrix: 33694 7461</p>
<p>Loading data of D1N â€¦</p>
<p>Dimensions of final gene expression matrix: 33694 5535</p>
<p>Loading data of D2N â€¦</p>
<p>Dimensions of final gene expression matrix: 33694 5145</p>
<pre class="r"><code>combined_obj &lt;- merge(seurat_lst[[1]], 
                      c(seurat_lst[[2]], seurat_lst[[3]], seurat_lst[[4]]),
                      add.cell.ids = filename_tb$experiment,
                      project = &quot;T_cells_all_merged&quot;)</code></pre>
<p>Dimensions of the merged gene expression matrix:</p>
<pre class="r"><code>paste0(&quot;Genes: &quot;, dim(combined_obj)[1])
paste0(&quot;Cells: &quot;, dim(combined_obj)[2])</code></pre>
<pre><code>[1] &quot;Genes: 33694&quot;
[1] &quot;Cells: 25094&quot;</code></pre>
</div>
<div id="quality-control" class="section level2" number="2.2">
<h2><span class="header-section-number">2.2</span> Quality control</h2>
<p>Next, Seurat is used to filter cells that contain &lt; 500 expressed
genes or more than 10% of total read counts from mitochondria genes.</p>
<p>The violin plots show the distributions of unique UMI count, library
size, and mitochondria gene percentage in cells from each batch.</p>
<pre class="r"><code>MT_genes &lt;- feature.names %&gt;% filter(startsWith(V2, &quot;MT-&quot;)) %&gt;% pull(V1)
combined_obj[[&#39;percent_mt&#39;]] &lt;- PercentageFeatureSet(combined_obj, 
                                                     features = MT_genes)
combined_obj &lt;- subset(combined_obj, 
                       subset = percent_mt &lt; 10 &amp; nFeature_RNA &gt; 500)
VlnPlot(combined_obj, 
        features = c(&#39;nFeature_RNA&#39;, &#39;nCount_RNA&#39;, &#39;percent_mt&#39;), 
        pt.size = 0)</code></pre>
<p><img src="preprocess_and_gsfa_TCells_files/figure-html/seurat%20QC-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Dimensions of the merged gene expression matrix after QC:</p>
<pre class="r"><code>paste0(&quot;Genes: &quot;, dim(combined_obj)[1])
paste0(&quot;Cells: &quot;, dim(combined_obj)[2])</code></pre>
<pre><code>[1] &quot;Genes: 33694&quot;
[1] &quot;Cells: 24955&quot;</code></pre>
The number of cells in each donor/stimulation status:
<div style="border: 1px solid #ddd; padding: 5px; ">
<table class="table" style="margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:left;">
status
</th>
<th style="text-align:right;">
number
</th>
<th style="text-align:left;">
description
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
TCells_D1N
</td>
<td style="text-align:right;">
5533
</td>
<td style="text-align:left;">
donor1 unstimulated
</td>
</tr>
<tr>
<td style="text-align:left;">
TCells_D1S
</td>
<td style="text-align:right;">
6843
</td>
<td style="text-align:left;">
donor1 stimulated
</td>
</tr>
<tr>
<td style="text-align:left;">
TCells_D2N
</td>
<td style="text-align:right;">
5144
</td>
<td style="text-align:left;">
donor2 unstimulated
</td>
</tr>
<tr>
<td style="text-align:left;">
TCells_D2S
</td>
<td style="text-align:right;">
7435
</td>
<td style="text-align:left;">
donor2 stimulated
</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="deviance-residual-transformation" class="section level2"
number="2.3">
<h2><span class="header-section-number">2.3</span> Deviance residual
transformation</h2>
<p>To accommodate the application of GSFA, we adopt the transformation
proposed in <a
href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1861-6">Townes
et al.</a>, and convert the scRNA-seq count data into deviance
residuals, a continuous quantity analogous to z-scores that
approximately follows a normal distribution.</p>
<p>The deviance residual transformation overcomes some problems of the
commonly used log transformation of read counts, and has been shown to
improve downstream analyses, such as feature selection and
clustering.</p>
<p>In the following, we used the
<code>deviance_residual_transform</code> function in <code>GSFA</code>
to perform the tranformation.<br />
<strong>Due to the size of the data, this process can take up to 1.1
hours, and should preferably be run in an R script
separately.</strong></p>
<pre class="r"><code>dev_res &lt;- deviance_residual_transform(t(as.matrix(combined_obj@assays$RNA@counts)))</code></pre>
</div>
<div id="feature-selection" class="section level2" number="2.4">
<h2><span class="header-section-number">2.4</span> Feature
selection</h2>
<p>Genes with constant expression across cells are not informative and
will have a deviance of 0, while genes that vary across cells in
expression will have a larger deviance.</p>
<p>Therefore, one can pick the genes with high deviance as an
alternative to selecting highly variable genes, with the advantage that
the selection is not sensitive to normalization.</p>
<p>Here we select the top 6000 genes ranked by deviance statistics, and
downsize the gene expression matrix.</p>
<pre class="r"><code>top_gene_index &lt;- select_top_devres_genes(dev_res, num_top_genes = 6000)
dev_res_filtered &lt;- dev_res[, top_gene_index]</code></pre>
</div>
<div id="covariate-removal" class="section level2" number="2.5">
<h2><span class="header-section-number">2.5</span> Covariate
removal</h2>
<p>We further regress out the differences in unique UMI count, library
size, and percentage of mitochondrial gene expression. As in the
original study, we choose not to correct for donor batch or stimulation
status as they contain genuine biological differences.</p>
<pre class="r"><code>covariate_df &lt;- data.frame(lib_size = combined_obj$nCount_RNA,
                           umi_count = combined_obj$nFeature_RNA,
                           percent_mt = combined_obj$percent_mt)
dev_res_corrected &lt;- covariate_removal(dev_res_filtered, covariate_df)

scaled.gene_exp &lt;- scale(dev_res_corrected)</code></pre>
</div>
<div id="inputs-for-gsfa" class="section level2" number="2.6">
<h2><span class="header-section-number">2.6</span> Inputs for GSFA</h2>
<p>The corrected and scaled gene expression matrix (24955 cells by 6000
genes) will be used as input for GSFA input argument <code>Y</code>. We
annonate this matrix with cell and gene names so we can keep track.</p>
<pre class="r"><code>sample_names &lt;- colnames(combined_obj@assays$RNA@counts)
gene_names &lt;- rownames(combined_obj@assays$RNA@counts)
rownames(scaled.gene_exp) &lt;- sample_names
colnames(scaled.gene_exp) &lt;- gene_names[top_gene_index]</code></pre>
<p>In addition, we have a cell by perturbation matrix containing
gene-level perturbation conditions across cells for GSFA input argument
<code>G</code>:</p>
<pre class="r"><code>G_mat &lt;- combined_obj@meta.data[, 4:24]
G_mat &lt;- as.matrix(G_mat)</code></pre>
<p>Here is the number of cells under each gene-level perturbation:</p>
<pre class="r"><code>num_cells &lt;- colSums(G_mat)
num_cells_df &lt;- data.frame(locus = names(num_cells),
                           count = num_cells)
ggplot(data = num_cells_df, aes(x=locus, y=count)) +
  geom_bar(stat = &quot;identity&quot;, width = 0.6) +
  labs(x = &quot;Perturbation&quot;,
       y = &quot;Number of cells&quot;) +
  theme_bw() + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size =11),
        axis.text = element_text(size = 12),
        axis.title = element_text(size = 14),
        panel.grid.minor = element_blank())</code></pre>
<p><img src="preprocess_and_gsfa_TCells_files/figure-html/num_cells-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Last but not least, since we hypothesize that perturbations may have
different effects in stimulated and unstimulated cells, we would like to
apply our modified two-group GSFA model on the data, so that the
perturbation effects can be estimated for cells under different
stimulation conditions separately.</p>
<p>Hence, we also need the cell group information as GSFA input argument
<code>group</code> Here we assign all unstimulated cells to group 0, and
all stimulated cells to group 1.</p>
<pre class="r"><code># Cell group info:
sample_group &lt;- combined_obj$orig.ident
sample_group &lt;- (sample_group %in% c(&quot;TCells_D1S&quot;, &quot;TCells_D2S&quot;)) * 1</code></pre>
<p>The number of cells in each group:</p>
<pre class="r"><code>table(sample_group)</code></pre>
<pre><code>sample_group
    0     1 
10677 14278 </code></pre>
</div>
</div>
<div id="gsfa" class="section level1" number="3">
<h1><span class="header-section-number">3</span> GSFA</h1>
<p>Now that we have all the inputs we need
(<code>scaled.gene_exp</code>, <code>G_mat</code>,
<code>sample_group</code>), we can perform GSFA (guided sparse factor
analysis) on the data.</p>
<p>We use the modified GSFA model with two cell groups (implemented in
<code>fit_gsfa_multivar_2groups()</code>), stratifying all cells by
their stimulation states (unstimulated:0, stimulated:1). We specify 20
factors initialized from truncated SVD and run Gibbs sampling for 4000
iterations in total, with the posterior mean estimates computed over the
last 1000 iterations of samples.</p>
<p>(Due to the size of data, this was performed in two segments, each
with 2000 Gibbs sampling iterations. More options for running GSFA are
available in the <code>R/run_gsfa_TCells_2groups.R</code> script.)</p>
<p><strong>This process is both time and memory intensive, and should be
run in an R script separately, preferably on a high performance
computing cluster that can guarantee 50GB memory and 6 (typically ~5.5)
hours of runtime without interruption for each segment of
run.</strong></p>
<pre class="r"><code># First segment of run (2k iterations):
set.seed(92629)
fit0 &lt;- fit_gsfa_multivar_2groups(Y = scaled.gene_exp, G = G_mat, 
                                  group = sample_group, K = 20,
                                  prior_type = &quot;mixture_normal&quot;, 
                                  init.method = &quot;svd&quot;,
                                  niter = 2000, used_niter = 1000,
                                  verbose = T, return_samples = T)</code></pre>
<pre class="r"><code># Second segment of run (2k iterations):
set.seed(92629)
fit &lt;- fit_gsfa_multivar_2groups(Y = scaled.gene_exp, G = G_mat, 
                                 group = sample_group, fit0 = fit0,
                                 prior_type = &quot;mixture_normal&quot;, 
                                 init.method = &quot;svd&quot;,
                                 niter = 2000, used_niter = 1000,
                                 verbose = T, return_samples = T)</code></pre>
</div>
<div id="session-information" class="section level1" number="4">
<h1><span class="header-section-number">4</span> Session
information</h1>
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.2.0 (2022-04-22)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: CentOS Linux 7 (Core)

Matrix products: default
BLAS/LAPACK: /software/openblas-0.3.13-el7-x86_64/lib/libopenblas_haswellp-r0.3.13.so

locale:
 [1] LC_CTYPE=en_US.UTF-8 LC_NUMERIC=C         LC_TIME=C           
 [4] LC_COLLATE=C         LC_MONETARY=C        LC_MESSAGES=C       
 [7] LC_PAPER=C           LC_NAME=C            LC_ADDRESS=C        
[10] LC_TELEPHONE=C       LC_MEASUREMENT=C     LC_IDENTIFICATION=C 

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] GSFA_0.2.8         SeuratObject_4.1.3 Seurat_4.3.0       Matrix_1.5-3      
 [5] forcats_1.0.0      stringr_1.5.0      dplyr_1.1.0        purrr_1.0.1       
 [9] readr_2.1.4        tidyr_1.3.0        tibble_3.1.8       ggplot2_3.4.1     
[13] tidyverse_1.3.2    data.table_1.14.6 

loaded via a namespace (and not attached):
  [1] readxl_1.4.2           backports_1.4.1        systemfonts_1.0.4     
  [4] plyr_1.8.7             igraph_1.4.0           lazyeval_0.2.2        
  [7] sp_1.6-0               splines_4.2.0          listenv_0.8.0         
 [10] scattermore_0.8        digest_0.6.31          htmltools_0.5.4       
 [13] fansi_1.0.4            magrittr_2.0.3         tensor_1.5            
 [16] googlesheets4_1.0.1    cluster_2.1.3          ROCR_1.0-11           
 [19] tzdb_0.3.0             globals_0.15.0         modelr_0.1.10         
 [22] matrixStats_0.63.0     R.utils_2.12.2         svglite_2.1.0         
 [25] timechange_0.2.0       spatstat.sparse_3.0-0  colorspace_2.1-0      
 [28] rvest_1.0.3            ggrepel_0.9.3          haven_2.5.1           
 [31] xfun_0.37              crayon_1.5.2           jsonlite_1.8.4        
 [34] progressr_0.10.0       spatstat.data_3.0-0    survival_3.3-1        
 [37] zoo_1.8-10             glue_1.6.2             kableExtra_1.3.4      
 [40] polyclip_1.10-0        gtable_0.3.1           gargle_1.3.0          
 [43] webshot_0.5.3          leiden_0.4.2           future.apply_1.9.0    
 [46] abind_1.4-5            scales_1.2.1           DBI_1.1.3             
 [49] spatstat.random_3.1-3  miniUI_0.1.1.1         Rcpp_1.0.10           
 [52] viridisLite_0.4.1      xtable_1.8-4           reticulate_1.24       
 [55] htmlwidgets_1.6.1      httr_1.4.4             RColorBrewer_1.1-3    
 [58] ellipsis_0.3.2         ica_1.0-2              farver_2.1.1          
 [61] R.methodsS3_1.8.2      pkgconfig_2.0.3        sass_0.4.5            
 [64] uwot_0.1.14            dbplyr_2.3.0           deldir_1.0-6          
 [67] utf8_1.2.3             labeling_0.4.2         tidyselect_1.2.0      
 [70] rlang_1.0.6            reshape2_1.4.4         later_1.3.0           
 [73] munsell_0.5.0          cellranger_1.1.0       tools_4.2.0           
 [76] cachem_1.0.6           cli_3.6.0              generics_0.1.3        
 [79] broom_1.0.3            ggridges_0.5.3         evaluate_0.20         
 [82] fastmap_1.1.0          yaml_2.3.7             goftest_1.2-3         
 [85] knitr_1.42             fs_1.6.1               fitdistrplus_1.1-8    
 [88] pander_0.6.5           RANN_2.6.1             pbapply_1.5-0         
 [91] future_1.25.0          nlme_3.1-157           mime_0.12             
 [94] ggrastr_1.0.1          R.oo_1.25.0            xml2_1.3.3            
 [97] compiler_4.2.0         rstudioapi_0.14        beeswarm_0.4.0        
[100] plotly_4.10.0          png_0.1-8              spatstat.utils_3.0-1  
[103] reprex_2.0.2           bslib_0.4.2            stringi_1.7.12        
[106] highr_0.10             lattice_0.20-45        vctrs_0.5.2           
[109] pillar_1.8.1           lifecycle_1.0.3        spatstat.geom_3.0-6   
[112] lmtest_0.9-40          jquerylib_0.1.4        RcppAnnoy_0.0.19      
[115] cowplot_1.1.1          irlba_2.3.5            httpuv_1.6.5          
[118] patchwork_1.1.1        R6_2.5.1               promises_1.2.0.1      
[121] KernSmooth_2.23-20     gridExtra_2.3          vipor_0.4.5           
[124] parallelly_1.34.0      codetools_0.2-18       MASS_7.3-56           
[127] assertthat_0.2.1       withr_2.5.0            sctransform_0.3.5     
[130] parallel_4.2.0         hms_1.1.2              grid_4.2.0            
[133] rmarkdown_2.20         googledrive_2.0.0      Rtsne_0.16            
[136] spatstat.explore_3.0-6 shiny_1.7.1            lubridate_1.9.2       
[139] ggbeeswarm_0.7.1      </code></pre>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
